---
title: "TGFbeta expression modeling"
author: "Migdal"
date: "8/18/2021"
output: pdf_document
---

```{r include=FALSE}
knitr::opts_chunk$set(warning = FALSE, message = FALSE)

devtools::load_all()

doMC::registerDoMC(20L)

def.par <- par(no.readonly = TRUE)

data(
  "promoters_f5_core",
  "remap_promoters_f5",
  "chip_atlas_promoters_f5",
  "promoters_f5",
  package = "xcoredata"
)
```

# Gene expression modeling in context of TGFbeta treatment

We used time series data from TGFβ treatment experiment for the development
of our expression modeling pipeline. The data include a series of CAGE-seq experiments
conducted in two biotype contexts: *A-549*, *MDA-MB-231* at 0, 1, 2, 4, 6 and 24
hours after the treatment. The CAGE reads were mapped to human genome assembly
Hg38 as described previously [ref]. Subsequently, mapped expression tags were
quantified against FANTOM5 promoters set. We further consider expression tags
counts at **DPI** and **SYMBOL** levels. In DPI we include expression tags at
all FANTOM5 promoters, while in SYMBOL we consider expression tags only at
**CORE PROMOTERS** (TODO).

```{r}
exprss_files <- list(
  system.file("inst", "extdata", "TGFbeta_counts_symbol.rda",
              package = "xcore"),
  system.file("inst", "extdata", "TGFbeta_counts_dpi.rda",
              package = "xcore"))
for(e in exprss_files) load(e)
rm(e)
```

# Prepare counts for regression

The CAGE expression tags for each sample are filtered for lowly epressed
promtoers, normalized for the library size and transformed into counts per
million using edgeR R package. Next, CPM are log2 transformed with addition of pseudo count 1. We designate the base level samples, 0h after treatment in our
TGFβ data, from which we calculate per gene mean expression over all replicates,
further called basal expression level ($\mu$).

```{r}
# construct sample matrix
samples <- data.frame(
  samples_names = counts_symbol %>% colnames(),
  exp_fac = counts_symbol %>%
    colnames() %>%
    sub(pattern = "_R.", replacement = ""),
  stringsAsFactors = FALSE)

# construct design matrix
design_mat <- model.matrix(~ 0 + exp_fac, data = samples)
colnames(design_mat) <- sub( "exp_fac", "" , colnames(design_mat))
rownames(design_mat) <- samples$samples_names

designA <- design_mat[
  grepl("A_", rownames(design_mat)),
  grepl("A_", colnames(design_mat)),
  drop = FALSE
  ]

designM <- design_mat[
  grepl("M_", rownames(design_mat)),
  grepl("M_", colnames(design_mat)),
  drop = FALSE
  ]
```

```{r}
symbol2dpi <- stats::setNames(promoters_f5_core$name, promoters_f5_core$SYMBOL)

# A-549
## SYMBOL
mat <- counts_symbol[, grep("A_", colnames(counts_symbol))]
rownames(mat) <- symbol2dpi[rownames(mat)]
base_lvl <- "A_TGFb_00h"
dataA_symbol <- prepareCountsForRegression(
  counts = mat,
  base_lvl = base_lvl,
  design = designA,
  log2 = TRUE,
  pseudo_count = 1L)

## DPI
mat <- counts_dpi[, grep("A_", colnames(counts_dpi))]
base_lvl <- "A_TGFb_00h"
dataA_dpi <- prepareCountsForRegression(
  counts = mat,
  base_lvl = base_lvl,
  design = designA,
  log2 = TRUE,
  pseudo_count = 1L)

# MDA-MB-231
## SYMBOL
mat <- counts_symbol[, grep("M_", colnames(counts_symbol))]
rownames(mat) <- symbol2dpi[rownames(mat)]
base_lvl <- "M_TGFb_00h"
dataM_symbol <- prepareCountsForRegression(
  counts = mat,
  base_lvl = base_lvl,
  design = designM,
  log2 = TRUE,
  pseudo_count = 1L)

## DPI
mat <- counts_dpi[, grep("M_", colnames(counts_dpi))]
base_lvl <- "M_TGFb_00h"
dataM_dpi <- prepareCountsForRegression(
  counts = mat,
  base_lvl = base_lvl,
  design = designM,
  log2 = TRUE,
  pseudo_count = 1L)
```

# Load signatures

The molecular signatures were constructed by first, extending FANTOM5 promoters
by 500bp in both direction creating a promoter regions set. Next, various
molecular signatures are intersected with promoter regions yielding a binary
molecular signature, where 1 indicates presence of a signature in the promoter
and 0 indicates it's absence. Binary molecular signatures were constructed for
two large ChIP-seq databases, namely ReMap2020 and ChIP-Atlas from where we
downloaded Hg38 peaks sets for all available transcription factors. For
ReMap2020 no further processing of peaks was applied, for ChIP-Atlas we have
excluded some ambiguous experiment such as those labeled as "Biotin" or
"Epitope" tags. Obtained signatures, were further clustered and collapsed yielding
collapsed ReMap2020 and ChIP-Atlas binary signatures at different levels of
collapsing. We found that collapsing based on Pearson distance and signatures
collapsing based on 30% majority rule gave best results.Additionally, we
consider molecular signatures for transcription factors motifs based on two
sources Jaspar and SwissRegulon databases.

```{r}
molecular_signatures <- list()

molecular_signatures[["remap"]] <- remap_promoters_f5
molecular_signatures[["chip_atlas"]] <- chip_atlas_promoters_f5


# collapsed ReMap2020
molecular_signatures <- as.environment(molecular_signatures)
collapsed_sig_files <- list.files(
  path = system.file("inst",
                     "analysis",
                     "into_signatures",
                     "clustering_diff_distances",
                     package = "xcore"),
  pattern = ".*_promoters_pearson_.*_0.3.rda",
  full.names = TRUE)
for (sig in collapsed_sig_files) load(file = sig, envir = molecular_signatures)
molecular_signatures <- as.list(molecular_signatures)

# motifs signatures                 count = FALSE)
# Jaspar
jaspar_input <-
    system.file("inst", "extdata", "JASPAR2020_hg38.promoters_f5.4kb.bed.gz", package = "xcore")
jaspar <- rtracklayer::import(jaspar_input)
molecular_signatures[["jaspar"]] <-
  getInteractionMatrix(a = promoters_f5,
                       b = jaspar,
                       ext = 500,
                       count = FALSE)
# SwissRegulon
swissregulon_input <- system.file("inst", "extdata", "hg38_sites_v1.gff.gz", package = "xcore")
swissregulon <- rtracklayer::import(swissregulon_input)
swissregulon$name <- swissregulon$Motif
molecular_signatures[["swissregulon"]] <-
  getInteractionMatrix(a = promoters_f5,
                       b = swissregulon,
                       ext = 500,
                       count = FALSE)
rm(jaspar, swissregulon)

# randomized ReMap2020, ChIP-Atlas
set.seed(12345432) # TDO might need to recompute for randomized
molecular_signatures[["remap_randomized"]] <-
  remap_promoters_f5[sample(nrow(remap_promoters_f5)), ]
rownames(molecular_signatures[["remap_randomized"]]) <- rownames(remap_promoters_f5)
```

# Combine expression with signatures

Here, we construct a `MultiAssayExperiment` as a container for expression and
signatures data. Additionally, we filter the molecular signatures such that only
those overlapping at least 5% and at most 95% of analyzed promoters are included.

```{r}
# A-549
## SYMBOL
molecular_signatures[["mae"]] <- dataA_symbol
dataA_symbol <- do.call(addSignatures, molecular_signatures)
dataA_symbol <- filterSignatures(dataA_symbol)
## DPI
molecular_signatures[["mae"]] <- dataA_dpi
dataA_dpi <- do.call(addSignatures, molecular_signatures)
dataA_dpi <- filterSignatures(dataA_dpi)

# MDA-MB-231
## SYMBOL
molecular_signatures[["mae"]] <- dataM_symbol
dataM_symbol <- do.call(addSignatures, molecular_signatures)
dataM_symbol <- filterSignatures(dataM_symbol)
## DPI
molecular_signatures[["mae"]] <- dataM_dpi
dataM_dpi <- do.call(addSignatures, molecular_signatures)
dataM_dpi <- filterSignatures(dataM_dpi)

rm(
  molecular_signatures,
  remap_promoters_f5,
  chip_atlas_promoters_f5,
  counts_dpi,
  counts_symbol,
  mat,
  promoters_f5,
  promoters_f5_core
)
gc()
```

# Gene expression modeling

## Regression type selection

```{r eval=FALSE}
regression_type <- list(A_549 = list(), MDA_MB_231 = list())
mask_24h <- colnames(dataA_dpi[["Y"]]) %>% grepl(pattern = "24h")
xname <- "remap"

runCvGlmnetFun <- function(mae, xname, ymask, nfolds = 10) {
  alpha <- c("lasso" = 1, "elasticnet" = 0.5, "ridge" = 0)
  res <- foreach(type = c("lasso", "elasticnet", "ridge")) %:%
    foreach(y = iterators::iter(mae[["Y"]][, ymask])) %dopar%
      glmnet::cv.glmnet(
        x = mae[[xname]][, ymask],
        y = y,
        family = "gaussian",
        alpha = alpha[type],
        standarize = TRUE,
        offset = mae[["U"]],
        nfolds = nfolds
      )
  names(res) <- c("lasso", "elasticnet", "ridge")

  return(res)
}

estimateR2 <- function(mae, regression_res, xname = xname, ymask = mask_24h) {
  res <- foreach(nm = names(regression_res)) %dopar%
    {
      foreach(i = seq_len(4L), .combine = c) %dopar%
        estimateStat(
          x = mae[[xname]],
          y = mae[["Y"]][, ymask][, i],
          u = mae[["U"]],
          s = regression_res[[nm]][[i]]$lambda.min,
          statistic = rsq,
          alpha = switch (nm,
            "lasso" = 1,
            "elasticnet" = 0.5,
            "ridge" = 0
          ))
  }
  names(res) <- names(regression_res)

  return(res)
}

regression_type[["A_549"]] <-
  runCvGlmnetFun(mae = dataA_dpi, xname = xname, ymask = mask_24h)
A549 <-
  estimateR2(
    mae = dataA_dpi,
    regression_res = regression_type$A_549,
    xname = xname,
    ymask = mask_24h
  )

regression_type[["MDA_MB_231"]] <-
  runCvGlmnetFun(mae = dataM_dpi, xname = xname, ymask = mask_24h)
MDAMB231 <-
  estimateR2(
    mae = dataM_dpi,
    regression_res = regression_type$MDA_MB_231,
    xname = xname,
    ymask = mask_24h
  )

save(A549, MDAMB231, file = "regression_type_selection.rda")
```

```{r}
load("regression_type_selection.rda")
```

```{r}
plotRegressionTypeSelection <- function() {
  dev.hold()
  on.exit(dev.flush())
  op <- par(no.readonly = TRUE)
  on.exit(par(op), add = TRUE)

  # define layout
  graphics::layout(
    mat = matrix(
      data = c(1, 1, 2, 3),
      ncol = 2,
      byrow = TRUE
    ),
  heights = c(2, 7))

  # title
  par(mar = c(1.1, 1.1, 1.1, 1.1))
  plot.new()
  graphics::text(
    x = 0.5,
    y = 0.5,
    adj = c(0.5, 0.5),
    labels = expression(atop(
      "Regularized regression methods comparison",
      paste("CV ", R ^ 2, " across replicates, ReMap2020")
    )),
    cex = 1.5,
    font = 2
  )

  # boxplots
  par(
    mar = c(6.1, 5.1, 1.1, 2.1),
    font.main = 1,
    cex.main = 1.1,
    cex = 1.0
  )
  graphics::boxplot(
    A549,
    main = "A-549 24h DPI",
    ylab = expression(R ^ 2),
    xlab = "",
    ylim = c(0.10, 0.36),
    las = 2
  )
  # graphics::stripchart(
  #   A549,
  #   method = "jitter",
  #   pch = 20,
  #   vertical = TRUE,
  #   add = TRUE
  # )

  graphics::boxplot(
    MDAMB231,
    main = "MDA-MB-231 24h DPI",
    ylab = expression(R ^ 2),
    xlab = "",
    ylim = c(0.10, 0.36),
    las = 2
  )
  # graphics::stripchart(
  #   MDAMB231,
  #   method = "jitter",
  #   pch = 20,
  #   vertical = TRUE,
  #   add = TRUE
  # )

  invisible()
}

plotRegressionTypeSelection()
```

## Ridge regression

```{r}
load("regression.rda")
```

```{r eval=FALSE}
regression <- list(A_549 = list(), MDA_MB_231 = list())

# A-549
## SYMBOL
xnames <- setdiff(names(dataA_symbol), c("U", "Y"))
regression[["A_549"]][["SYMBOL"]] <- modelGeneExpression(
  mae = dataA_symbol,
  yname = "Y",
  uname = "U",
  xnames = xnames,
  pvalues = FALSE,
  precalcmodels = regression[["A_549"]][["SYMBOL"]][["regression_models"]]
  )
save(regression, file = "regression.rda")

## DPI
xnames <- setdiff(names(dataA_dpi), c("U", "Y"))
regression[["A_549"]][["DPI"]] <- modelGeneExpression(
  mae = dataA_dpi,
  yname = "Y",
  uname = "U",
  xnames = xnames,
  pvalues = FALSE,
  precalcmodels = regression[["A_549"]][["DPI"]][["regression_models"]])
save(regression, file = "regression.rda")

# MDA-MB-231
## SYMBOL
xnames <- setdiff(names(dataM_symbol), c("U", "Y"))
regression[["MDA_MB_231"]][["SYMBOL"]] <- modelGeneExpression(
  mae = dataM_symbol,
  yname = "Y",
  uname = "U",
  xnames = xnames,
  pvalues = FALSE,
  precalcmodels = regression[["MDA_MB_231"]][["SYMBOL"]][["regression_models"]])
save(regression, file = "regression.rda")

## DPI
xnames <- setdiff(names(dataM_dpi), c("U", "Y"))
regression[["MDA_MB_231"]][["DPI"]] <- modelGeneExpression(
  mae = dataM_dpi,
  yname = "Y",
  uname = "U",
  xnames = xnames,
  pvalues = FALSE,
  precalcmodels = regression[["MDA_MB_231"]][["DPI"]][["regression_models"]])
save(regression, file = "regression.rda")
```

# R2 for different molecular signatures, 0 vs 24 hours

```{r eval=FALSE}
## A-549 24h
j <- grepl("24h", colnames(dataA_dpi[["Y"]]))
xnames <- setdiff(names(dataA_dpi), c("U", "Y"))
rsqA <- list(symbol = list(), dpi = list())
for (nm in xnames) {
  # symbol
  stat <-
    foreach(i = seq_len(sum(j)), .combine = c) %dopar%
    estimateStat(
      x = dataA_symbol[[nm]],
      y = dataA_symbol[["Y"]][, j][, i],
      u = dataA_symbol[["U"]],
      s = regression[["A_549"]][["SYMBOL"]][["regression_models"]][[nm]][j][[i]]$lambda.min,
      statistic = rsq)
  rsqA[["symbol"]][[nm]] <- stat

  # dpi
  stat <-
    foreach(i = seq_len(sum(j)), .combine = c) %dopar%
    estimateStat(
      x = dataA_dpi[[nm]],
      y = dataA_dpi[["Y"]][, j][, i],
      u = dataA_dpi[["U"]],
      s = regression[["A_549"]][["DPI"]][["regression_models"]][[nm]][j][[i]]$lambda.min,
      statistic = rsq)
  rsqA[["dpi"]][[nm]] <- stat
}

save(rsqA, file = "rsqA.rda")

## MDA-MB-231 24h
j <- grepl("24h", colnames(dataM_dpi[["Y"]]))
xnames <- setdiff(names(dataM_dpi), c("U", "Y"))
rsqM <- list(symbol = list(), dpi = list())
for (nm in xnames) {
  # symbol
  stat <-
    foreach(i = seq_len(sum(j)), .combine = c) %dopar%
    estimateStat(
      x = dataM_symbol[[nm]],
      y = dataM_symbol[["Y"]][, j][, i],
      u = dataM_symbol[["U"]],
      s = regression[["MDA_MB_231"]][["SYMBOL"]][["regression_models"]][[nm]][j][[i]]$lambda.min,
      statistic = rsq)
  rsqM[["symbol"]][[nm]] <- stat

  # dpi
  stat <-
    foreach(i = seq_len(sum(j)), .combine = c) %dopar%
    estimateStat(
      x = dataM_dpi[[nm]],
      y = dataM_dpi[["Y"]][, j][, i],
      u = dataM_dpi[["U"]],
      s = regression[["MDA_MB_231"]][["DPI"]][["regression_models"]][[nm]][j][[i]]$lambda.min,
      statistic = rsq)
  rsqM[["dpi"]][[nm]] <- stat
}

save(rsqM, file = "rsqM.rda")
```

```{r}
load("rsqA.rda")
load("rsqM.rda")
```

```{r}
plotMolecularSignaturesSetComparison <- function() {
  cex.axis <- 0.8

  dev.hold()
  on.exit(dev.flush())
  op <- par(no.readonly = TRUE)
  on.exit(par(op), add = TRUE)

  xsel <- c("remap", "chip_atlas", "jaspar", "swissregulon", "remap_randomized")
  xlabs <- c(
    "ReMap2020",
    "ChIP-Atlas",
    "Jaspar",
    "SwissRegulon",
    "randomized\nReMap2020"
  )

  # define layout
  graphics::layout(mat = matrix(
    data = c(1, 2, 1, 3),
    nrow = 2,
    ncol = 2
  ),
  heights = c(2, 7))

  # title
  par(mar = c(0.5, 1.1, 1.1, 1.1))
  plot.new()
  graphics::text(
    x = 0.5,
    y = 0.5,
    adj = c(0.5, 0.5),
    labels = expression(atop(
      "Molecular signatures sets comparison",
      paste("CV ", R^2, " across replicates, 24h DPI")
    )),
    cex = 1.8,
    font = 2
  )

  # boxplots
  par(mar = c(9.1, 5.1, 2.1, 1.1), font.main = 1, cex.main = 1.1, cex = 1.0)
  graphics::boxplot(
    x = rsqA$dpi[xsel],
    names = xlabs,
    main = "A-549",
    ylab = expression(R ^ 2),
    xlab = "",
    ylim = c(0, 0.4),
    las = 2,
    cex.axis = cex.axis
  )
  graphics::boxplot(
    x = rsqM$dpi[xsel],
    names = xlabs,
    main = "MDA-MB-549",
    ylab = expression(R ^ 2),
    xlab = "",
    ylim = c(0, 0.4),
    las = 2,
    cex.axis = cex.axis
  )

  invisible()
}

plotMolecularSignaturesSetComparison()
```

### Collapsing leads to loss of model accuracy
#### A-549 24h

```{r}
plotScatterPlotWithErrorBars <- function(rsq, ord, xsize, title_suffix = "A-549 24h") {
  dev.hold()
  on.exit(dev.flush())
  op <- par(no.readonly = TRUE)
  on.exit(par(op), add = TRUE)

  par(mar = c(10, 4, 4, 2) + 0.1)
  plot(
    x = NULL,
    y = NULL,
    xlim = c(max(xsize), min(xsize)),
    ylim = c(0, max(unlist(rsq$dpi[ord]))),
    xlab = "",
    ylab = expression(R ^ 2),
    xaxt = "n",
    main = expression(paste("CV ", R ^ 2, " across replicates,")),
    sub = title_suffix
  )
  axis(
    side = 1,
    at = xsize,
    labels = ord %>%
      sub(pattern = ".*_promoters_pearson_", replacement = "") %>%
      sub(pattern = "_0.3$", replacement = "") %>%
      paste0(" (", xsize, ")"),
    las = 2
  )

  symbol_means <-
    vapply(ord, function(x)
      mean(rsq[["symbol"]][[x]]), FUN.VALUE = numeric(1L))
  symbol_sd <-
    vapply(ord, function(x)
      sd(rsq[["symbol"]][[x]]), FUN.VALUE = numeric(1L))
  points(x = xsize - 30,
         y = symbol_means,
         pch = 16)
  arrows(
    x0 = xsize - 30,
    y0 = symbol_means - symbol_sd,
    y1 = symbol_means + symbol_sd,
    length = 0.05,
    angle = 90,
    code = 3
  )

  dpi_means <-
    vapply(ord, function(x)
      mean(rsq[["dpi"]][[x]]), FUN.VALUE = numeric(1L))
  dpi_sd <-
    vapply(ord, function(x)
      sd(rsq[["dpi"]][[x]]), FUN.VALUE = numeric(1L))
  points(x = xsize + 30,
         y = dpi_means,
         pch = 1)
  arrows(
    x0 = xsize + 30,
    y0 = dpi_means - dpi_sd,
    y1 = dpi_means + dpi_sd,
    length = 0.05,
    angle = 90,
    code = 3
  )
  legend("topright",
         legend = c("dpi", "symbol"),
         pch = c(1, 16))
}


plotBoxplotWithTwoGroups <-
  function(rsq, ord, title_suffix = "A-549 24h") {
    df <- rbind(
      data.frame(
        value = unlist(rsq$dpi[ord]),
        xname = ordered(
          x = rep(names(rsq$dpi[ord]), each = length(rsq$dpi$remap)) %>%
            sub(pattern = ".*_promoters_pearson_", replacement = "") %>%
            sub(pattern = "_0.3$", replacement = ""),
          levels = ord %>%
            sub(pattern = ".*_promoters_pearson_", replacement = "") %>%
            sub(pattern = "_0.3$", replacement = "")
        ),
        type = "dpi",
        stringsAsFactors = FALSE,
        row.names = NULL
      ),
      data.frame(
        value = unlist(rsq$symbol[ord]),
        xname = ordered(
          x = rep(names(rsq$symbol[ord]), each = length(rsq$symbol$remap)) %>%
            sub(pattern = ".*_promoters_pearson_", replacement = "") %>%
            sub(pattern = "_0.3$", replacement = ""),
          levels = ord %>%
            sub(pattern = ".*_promoters_pearson_", replacement = "") %>%
            sub(pattern = "_0.3$", replacement = "")
        ),
        type = "symbol",
        stringsAsFactors = FALSE,
        row.names = NULL
      )
    )

    ggplot2::ggplot(df, ggplot2::aes(x = xname, y = value, fill = type)) +
      ggplot2::geom_boxplot() +
      ggplot2::ylim(0, 0.35) +
      ggplot2::xlab("molecular signature") +
      ggplot2::ylab(expression(R ^ 2)) +
      ggplot2::ggtitle(expression(paste(
        "CV ", R ^ 2, " across replicates,"
      )),
        subtitle = title_suffix) +
      ggplot2::theme_bw(base_size = 14)
  }
```

```{r}
ord_remap <-
  c(
    "remap",
    "remap_promoters_pearson_0.05_0.3",
    "remap_promoters_pearson_0.09_0.3",
    "remap_promoters_pearson_0.13_0.3",
    "remap_promoters_pearson_0.17_0.3",
    "remap_promoters_pearson_0.21_0.3",
    "remap_promoters_pearson_0.5_0.3",
    "remap_promoters_pearson_0.9_0.3",
    "remap_promoters_pearson_1.3_0.3",
    "remap_promoters_pearson_1.7_0.3"
  )

xsize_remap <- vapply(ord_remap, function(x) ncol(dataA_dpi[[x]]), FUN.VALUE = numeric(1L))
plotScatterPlotWithErrorBars(rsqA, ord_remap, xsize_remap, title_suffix = "A-549 ReMap2020 24h")
plotBoxplotWithTwoGroups(rsqA, ord_remap, title_suffix = "A-549 ReMap2020 24h")

xsize_remap <- vapply(ord_remap, function(x) ncol(dataM_dpi[[x]]), FUN.VALUE = numeric(1L))
plotScatterPlotWithErrorBars(rsqM, ord_remap, xsize_remap, title_suffix = "MDA-MB-231 ReMap2020 24h")
plotBoxplotWithTwoGroups(rsqM, ord_remap, title_suffix = "MDA-MB-231 ReMap2020 24h")
```

```{r}
ord_chip_atlas <-
  c(
    "chip_atlas",
    "chip_atlas_promoters_pearson_0.05_0.3",
    "chip_atlas_promoters_pearson_0.09_0.3",
    "chip_atlas_promoters_pearson_0.13_0.3",
    "chip_atlas_promoters_pearson_0.17_0.3",
    "chip_atlas_promoters_pearson_0.21_0.3",
    "chip_atlas_promoters_pearson_0.5_0.3",
    "chip_atlas_promoters_pearson_0.9_0.3",
    "chip_atlas_promoters_pearson_1.3_0.3",
    "chip_atlas_promoters_pearson_1.7_0.3"
  )

xsize_chip_atlas <- vapply(ord_chip_atlas, function(x) ncol(dataA_dpi[[x]]), FUN.VALUE = numeric(1L))
plotScatterPlotWithErrorBars(rsqA, ord_chip_atlas, xsize_chip_atlas, title_suffix = "A-549 ChIP-Atlas 24h")
plotBoxplotWithTwoGroups(rsqA, ord_chip_atlas, title_suffix = "A-549 ChIP-Atlas 24h")

xsize_chip_atlas <- vapply(ord_chip_atlas, function(x) ncol(dataM_dpi[[x]]), FUN.VALUE = numeric(1L))
plotScatterPlotWithErrorBars(rsqM, ord_chip_atlas, xsize_chip_atlas, title_suffix = "MDA-MB-231 ChIP-Atlas 24h")
plotBoxplotWithTwoGroups(rsqM, ord_chip_atlas, title_suffix = "MDA-MB-231 ChIP-Atlas 24h")
```

# Significant signatures selection

```{r eval=FALSE}
top_regression <- list(A_549 = list(), MDA_MB_231 = list())
xnames <- c("remap", "chip_atlas", "jaspar", "swissregulon")

top_regression[["A_549"]][["DPI"]] <- modelGeneExpression(
  mae = dataA_dpi,
  yname = "Y",
  uname = "U",
  xnames = xnames,
  pvalues = TRUE,
  precalcmodels = regression[["A_549"]][["DPI"]][["regression_models"]][xnames])
save(top_regression, file = "top_regression.rda")

top_regression[["MDA_MB_231"]][["DPI"]] <- modelGeneExpression(
  mae = dataM_dpi,
  yname = "Y",
  uname = "U",
  xnames = xnames,
  pvalues = TRUE,
  precalcmodels = regression[["MDA_MB_231"]][["DPI"]][["regression_models"]][xnames])
save(top_regression, file = "top_regression.rda")
```

```{r}
load("top_regression.rda")
```

# Stability of best signatures selection between replicates
## ReMap2020 24h top signatures across replicates

```{r}
selA <- c("A_TGFb_24h_R1", "A_TGFb_24h_R2", "A_TGFb_24h_R3", "A_TGFb_24h_R4")
signatures_mat_24_A549 <- do.call(
  what = cbind,
  args = lapply(
    X = selA,
    FUN = function(x) top_regression$A_549$DPI$pvalues$remap[[x]][["coef"]]
  )
)
colnames(signatures_mat_24_A549) <- c("R1", "R2", "R3", "R4")

selM <- c("M_TGFb_24h_R1", "M_TGFb_24h_R2", "M_TGFb_24h_R3", "M_TGFb_24h_R4")
signatures_mat_24_MDA_MB_231 <- do.call(
  what = cbind,
  args = lapply(
    X = selM,
    FUN = function(x) top_regression$MDA_MB_231$DPI$pvalues$remap[[x]][["coef"]]
  )
)
colnames(signatures_mat_24_MDA_MB_231) <- c("R1", "R2", "R3", "R4")

topA_549 <- applyOverDFList(
  list_of_df = top_regression$A_549$DPI$pvalues$remap[selA],
  col_name = "pval",
  fun = fisherMethod,
  groups = factor(setNames(rep(1, 4), names(top_regression$A_549$DPI$pvalues$remap[selA])))
) %>%
  drop() %>%
  sort() %>%
  head(25) %>%
  names()

topMDA_MB_231 <- applyOverDFList(
  list_of_df = top_regression$MDA_MB_231$DPI$pvalues$remap[selM],
  col_name = "pval",
  fun = fisherMethod,
  groups = factor(setNames(rep(1, 4), names(top_regression$MDA_MB_231$DPI$pvalues$remap[selM])))
) %>%
  drop() %>%
  sort() %>%
  head(25) %>%
  names()
```

```{r}
# recommended device size  2410 x 1082
graphics::layout(
  mat = matrix(
    data = c(1, 2, 3,
             0, 2, 3),
    nrow = 2,
    ncol = 3,
    byrow = TRUE),
  heights = c(1, 4),
  widths = c(1, 6, 6)
)

breaks <- seq(from = -0.08, to = 0.08, length.out = 51)

plotHeatmapScaleBar(breaks, main = "activity", mar = c(1.1, 4.1, 3.1, 2.1))

mat <- signatures_mat_24_A549[intersect(topA_549, topMDA_MB_231), ]
plotHeatmap(
  mat,
  breaks,
  main = "A-549",
  clust_rows = FALSE,
  cex.main = 1.3,
  cex.axis = 1.1,
  mar = c(6.1, 0.1, 4.1, 24.1)
)

mat <- signatures_mat_24_MDA_MB_231[intersect(topA_549, topMDA_MB_231), ]
plotHeatmap(
  mat,
  breaks,
  main = "MDA-MB-231",
  clust_rows = FALSE,
  cex.main = 1.3,
  cex.axis = 1.1,
  mar = c(6.1, 0.1, 4.1, 24.1)
)
```

## ChIP-Atlas 24h top signatures across replicates

```{r}
selA <- c("A_TGFb_24h_R1", "A_TGFb_24h_R2", "A_TGFb_24h_R3", "A_TGFb_24h_R4")
signatures_mat_24_A549 <- do.call(
  what = cbind,
  args = lapply(
    X = selA,
    FUN = function(x) top_regression$A_549$DPI$pvalues$chip_atlas[[x]][["coef"]]
  )
)
colnames(signatures_mat_24_A549) <- c("R1", "R2", "R3", "R4")

selM <- c("M_TGFb_24h_R1", "M_TGFb_24h_R2", "M_TGFb_24h_R3", "M_TGFb_24h_R4")
signatures_mat_24_MDA_MB_231 <- do.call(
  what = cbind,
  args = lapply(
    X = selM,
    FUN = function(x) top_regression$MDA_MB_231$DPI$pvalues$chip_atlas[[x]][["coef"]]
  )
)
colnames(signatures_mat_24_MDA_MB_231) <- c("R1", "R2", "R3", "R4")

topA_549 <- applyOverDFList(
  list_of_df = top_regression$A_549$DPI$pvalues$chip_atlas[selA],
  col_name = "pval",
  fun = fisherMethod,
  groups = factor(setNames(rep(1, 4), names(top_regression$A_549$DPI$pvalues$remap[selA])))
) %>%
  drop() %>%
  sort() %>%
  head(25) %>%
  names()

topMDA_MB_231 <- applyOverDFList(
  list_of_df = top_regression$MDA_MB_231$DPI$pvalues$chip_atlas[selM],
  col_name = "pval",
  fun = fisherMethod,
  groups = factor(setNames(rep(1, 4), names(top_regression$MDA_MB_231$DPI$pvalues$remap[selM])))
) %>%
  drop() %>%
  sort() %>%
  head(25) %>%
  names()
```

```{r}
graphics::layout(
  mat = matrix(
    data = c(1, 2, 3,
             0, 2, 3),
    nrow = 2,
    ncol = 3,
    byrow = TRUE),
  heights = c(1, 4),
  widths = c(1, 6, 6)
)

breaks <- seq(from = -0.08, to = 0.08, length.out = 51)

plotHeatmapScaleBar(breaks, main = "activity", mar = c(1.1, 4.1, 3.1, 2.1))

mat <- signatures_mat_24_A549[intersect(topA_549, topMDA_MB_231), ]
plotHeatmap(
  mat,
  breaks,
  main = "A-549",
  clust_rows = FALSE,
  cex.main = 1.3,
  cex.axis = 1.1,
  mar = c(6.1, 0.1, 4.1, 24.1)
)

mat <- signatures_mat_24_MDA_MB_231[intersect(topA_549, topMDA_MB_231), ]
plotHeatmap(
  mat,
  breaks,
  main = "MDA-MB-231",
  clust_rows = FALSE,
  cex.main = 1.3,
  cex.axis = 1.1,
  mar = c(6.1, 0.1, 4.1, 24.1)
)
```

# Top scoring signatures in TGFbeta - heatmaps

```{r}
plotTopSignaturesHeatmap <- function(signature = "remap") {
  dev.hold()
  on.exit(dev.flush())
  op <- par(no.readonly = TRUE)
  on.exit(par(op), add = TRUE)

  top10A_549 <- apply(top_regression$A_549$DPI$zscore_avg[[signature]], 1, stoufferZMethod) %>%
    abs() %>%
    sort(decreasing = TRUE) %>%
    head(10) %>%
    names()

  top10MDA_MB_231 <- apply(top_regression$MDA_MB_231$DPI$zscore_avg[[signature]], 1, stoufferZMethod) %>%
    abs() %>%
    sort(decreasing = TRUE) %>%
    head(10) %>%
    names()

  rows <- base::union(top10A_549, top10MDA_MB_231)

  am <- applyOverDFList(
      list_of_df = top_regression$A_549$DPI$pvalues[[signature]],
      col_name = "coef",
      fun = mean,
      groups = design2factor(metadata(dataA_dpi)$design)) %>%
    subsetWithMissing(rows = rows)

  mm <- applyOverDFList(
      list_of_df = top_regression$MDA_MB_231$DPI$pvalues[[signature]],
      col_name = "coef",
      fun = mean,
      groups = design2factor(metadata(dataM_dpi)$design)) %>%
    subsetWithMissing(rows = rows)

  mats_cb <- cbind(am, mm)
  i_na <- is.na(mats_cb)
  mats_cb[i_na] <- 0
  hc <- hclust(dist(mats_cb))
  ord <- order.dendrogram(as.dendrogram(hc))
  i <- rownames(am)[rev(ord)]

  graphics::layout(
    mat = matrix(
      data = c(1, 2, 3,
              0, 2, 3),
      nrow = 2,
      ncol = 3,
      byrow = TRUE),
    heights = c(1, 4),
    widths = c(1, 6, 6)
  )

  breaks <- seq(from = -0.08, to = 0.08, length.out = 51)

  plotHeatmapScaleBar(breaks, main = "activity", mar = c(1.1, 4.1, 3.1, 2.1))

  mat <- am[i,]
  colnames(mat) <- sub("._TGFb_", "", colnames(mat))
  plotHeatmap(
    mat,
    breaks,
    main = "A-549",
    clust_rows = FALSE,
    cex.main = 1.3,
    cex.axis = 1.1,
    mar = c(6.1, 0.1, 4.1, 24.1)
  )

  mat <- mm[i, ]
  colnames(mat) <- sub("._TGFb_", "", colnames(mat))
  plotHeatmap(
    mat,
    breaks,
    main = "MDA-MB-231",
    clust_rows = FALSE,
    cex.main = 1.3,
    cex.axis = 1.1,
   mar = c(6.1, 0.1, 4.1, 24.1)
  )

  invisible()
}
```

## ReMap2020

```{r}
plotTopSignaturesHeatmap("remap")
```

## ChIP-Atlas

```{r}
plotTopSignaturesHeatmap("chip_atlas")
```

## Jaspar

```{r}
plotTopSignaturesHeatmap("jaspar")
```

## SwissRegulon

```{r}
plotTopSignaturesHeatmap("swissregulon")
```

# Signature score vs # features 24h

```{r}
sig_th <- 1.96

# define layout
nf <- graphics::layout(
  mat = matrix(
    data = c(1, 2, 4, 5, 1, 3, 4, 6),
    nrow = 4,
    ncol = 2),
  heights = c(1, 4, 1, 4))

par(mar = c(1, 1, 1, 1))
plot.new()
text(x = 0.5,
     y = 0.5,
     labels = "A-549 24h DPI",
     cex = 1.5,
     font = 2)
par(mar = def.par$mar)

y <- top_regression$A_549$DPI$zscore_avg$remap[, "A_TGFb_24h"] %>% abs()
x <- colSums(dataA_dpi[["remap"]])[names(y)] %>% log10()
smoothScatter(
  x = x,
  y = y,
  xlab = "log10(signature coverage)",
  ylab = "abs(Z-score)",
  main = "ReMap2020",
  cex = 1.2)
abline(h = sig_th, col = "red", lty = 2)

y <- top_regression$A_549$DPI$zscore_avg$chip_atlas[, "A_TGFb_24h"] %>% abs()
x <- colSums(dataA_dpi[["chip_atlas"]])[names(y)] %>% log10()
smoothScatter(
  x = x,
  y = y,
  xlab = "log10(signature coverage)",
  ylab = "abs(Z-score)",
  main = "ChIP-Atlas",
  cex = 1.2)
abline(h = sig_th, col = "red", lty = 2)

par(mar = c(1, 1, 1, 1))
plot.new()
text(x = 0.5,
     y = 0.5,
     labels = "MDA-MB-231 24h DPI",
     cex = 1.5,
     font = 2)
par(mar = def.par$mar)

y <- top_regression$MDA_MB_231$DPI$zscore_avg$remap[, "M_TGFb_24h"] %>% abs()
x <- colSums(dataM_dpi[["remap"]])[names(y)] %>% log10()
smoothScatter(
  x = x,
  y = y,
  xlab = "log10(signature coverage)",
  ylab = "abs(Z-score)",
  main = "ReMap2020",
  cex = 1.2)
abline(h = sig_th, col = "red", lty = 2)

y <- top_regression$MDA_MB_231$DPI$zscore_avg$chip_atlas[, "M_TGFb_24h"] %>% abs()
x <- colSums(dataM_dpi[["chip_atlas"]])[names(y)] %>% log10()
smoothScatter(
  x = x,
  y = y,
  xlab = "log10(signature coverage)",
  ylab = "abs(Z-score)",
  main = "ChIP-Atlas",
  cex = 1.2)
abline(h = sig_th, col = "red", lty = 2)
```
