---
title: "FANTOM CAT CAGE peaks vs ChIP-seq data"
author: "Maciej Migda≈Ç"
output: pdf_document
params:
  cage_clusters: FANTOM_CAT.lv3_robust.CAGE_cluster.liftOver.hg38.bed.gz
  info_table: FANTOM_CAT.lv3_robust.info_table.ID_mapping.tsv.gz
  expression: FANTOM_CAT.expression_atlas.CAGE_cluster.lv1_raw.rle_cpm.tsv.gz
  gtf: FANTOM_CAT.lv3_robust.only_mRNA.gtf.gz
  main_string: only_mRNA
  genome: hg38
---

```{r include=FALSE}
knitr::opts_chunk$set(warning = FALSE, cache.lazy = FALSE)

devtools::load_all()
library(data.table)
```

```{r functions_data, cache=TRUE, echo=FALSE, message=FALSE}
# Internally used functions

# Function return summary as a string
summaryString <- function(...) {
    s <- summary(...)
    snm <- names(s)
    sval <- s %>% round(2) %>% as.character()
    len <- vapply(X = seq_along(snm), 
                  FUN = function(i) max(nchar(c(snm[i], sval[i]))) + 3L, 
                  FUN.VALUE = integer(1L))
    snm <- paste(vapply(
        X = seq_along(snm),
        FUN = function(i) stringr::str_pad(snm[i], len[i]),
        FUN.VALUE = character(1L)
        ), collapse = " ")
    sval <- paste(vapply(
        X = seq_along(sval),
        FUN = function(i) stringr::str_pad(sval[i], len[i] + 2),
        FUN.VALUE = character(1L)
        ), collapse = " ")
    paste0(snm, "\n", sval)
}

# Function calculates TF or Context coverage based on interaction matrix
# col to use for grouping
getCoverage <- function(mat, col = c(2, 3), sep = "\\.") {
    n <- stringr::str_count(colnames(mat)[1], sep)
    colnm <- colnames(mat) %>% 
        gsub(pattern = paste(c("(.*)", rep(c(sep, "(.*)"), n)), collapse = ""),
             replacement = paste0("\\", col))
    DelayedArray::colsum(x = mat, group = colnm %>% as.factor()) %>%
        `>`(1) %>%
        rowSums()
}

# Function plot coverage histogram
coverageHist<- function(x, breaks = 40, cex = 0.7, xlab = "", ...) {
    par(mar = c(5, 5, 6, 2))
    graphics::hist(x = x, 
                   breaks = breaks,
                   cex = 0.7,
                   xlab = xlab,
                   ...)
    mtext(text = summaryString(x), 
          side = 3,
          cex=0.6)
}

# Function plot two group box plot
blacklistBoxPlot <- function(coverage, cov_names, gr1, aname, bname, ...) {
    blk <- coverage[cov_names %in% gr1]
    nonblk <- coverage[! cov_names %in% gr1]
    aname <- paste0(aname, " (%i)")
    bname <- paste0(bname, " (%i)")

    graphics::boxplot(blk, nonblk,
                      names = c(sprintf(aname, length(blk)), 
                                sprintf(bname, length(nonblk))),
                      ...)
}
```

```{r static_data, cache=TRUE, echo=FALSE}
# fantom cat
cage_clusters <- data.table::fread(system.file("inst",
                                               "extdata",
                                               params$cage_clusters,
                                               package = "xcore"),
                                   header = FALSE)
colnames(cage_clusters) <- c("chr", "start", "end", "name", "score", "strand")
info_table <- data.table::fread(file = system.file("inst",
                                                   "extdata",
                                                   params$info_table,
                                                   package = "xcore"),
                                header = TRUE)
fantomcat <- info_table[cage_clusters,
                        .(chr, start, end, name, score, strand, geneID, transcriptID),
                        on = c("CAGEClusterID" = "name")]
# in some cases a signle promoter is assigned to multiple transcripts but same 
# genes eg. chr1:564571..564600,+ is assigned to ENST00000416931.1, 
# FTMT20300036855.1. Here we resolve this by droping duplicates.
fantomcat <- fantomcat[! duplicated(fantomcat$name), ]
rm(cage_clusters, info_table)
gc(verbose = FALSE)

# promoter expression
fantomcat_expression <-
    data.table::fread(
        file = system.file("inst",
                           "extdata",
                           params$expression,
                           package = "xcore"),
        header = TRUE
    )

fantomcat_expression <-
    data.table::melt(data = fantomcat_expression,
                     id.vars = "prmtrID")
fantomcat_expression <- fantomcat_expression[, .(
                         mean = mean(value, na.rm = TRUE),
                         min = min(value, na.rm = TRUE),
                         max = max(value, na.rm = TRUE),
                         sd = sd(value, na.rm = TRUE),
                         tau = fTau(log(value + 1))
                     ),
                     by = "prmtrID"]
gc()

# remap
remap_input <-
  system.file("inst",
              "extdata",
              "remap2020_all_macs2_hg38_v1_0.bed.gz",
              package = "xcore")
remap <- rtracklayer::import(remap_input)
remap$name <-
  gsub("\\s+", "", remap$name) # remove white spaces from features names

# chip-atlas
chip_atlas_file <-
  system.file(
    "inst",
    "extdata",
    "chip_atlas_hg38.Oth.ALL.05.AllAg.AllCell_SRX_only.bed.gz",
    package = "xcore"
  )
chip_atlas <- rtracklayer::import(chip_atlas_file)
# remove ambiguous records
ambiguous_tf <- c("5-hmC", "5-mC", "8-Hydroxydeoxyguanosine", "Biotin", "BMI",
                  "BrdU", "Cas9", "Cyclobutane", "MethylCap", "O-GlcNAc",
                  "Pan-acetyllysine", "pFM2", "RTA", "SVS-1", "EBNA1", "EBNA2",
                  "EBNA3", "EBV-ZEBRA", "AML1-ETO", "VSV-G", "MLL-AF4",
                  "MLL-AF6", "Hepatitis", "HIV", "KSHV", "MCPV", "Epitope tags",
                  "GFP")
ambiguous_srx <-
  read.delim(
    file = system.file("inst", "extdata", "experimentList_TF_hg38.txt", package = "xcore"),
    sep = "\t",
    header = FALSE,
    quote = "\"") %>%
  dplyr::filter(V4 %in% ambiguous_tf) %>% # V4 == TF column
  dplyr::pull(V1) # V1 == experiment ID
chip_atlas <- chip_atlas[! chip_atlas$name %in% ambiguous_srx, ]

sxr_meta_file <-
  system.file("inst", "extdata", "experimentList_TF_hg38.txt", package = "xcore")
sxr_meta <- read.delim(file = sxr_meta_file,
		      sep ="\t",
		      header = FALSE,
		      quote = "\"")
sxr_meta <- sxr_meta[sxr_meta$V1 %in% chip_atlas$name, ]
sxr_meta <- sxr_meta[, c(1, 4:7)]
colnames(sxr_meta) <- c("id", "TF", "origin", "cell", "description")
sxr_meta <- data.frame(apply(sxr_meta, 2, as.character), stringsAsFactors = FALSE)
sxr_name <- paste(sxr_meta$TF, sxr_meta$origin, sxr_meta$cell, sxr_meta$id, sep = "_")
sxr_name <- gsub(" " , ".", sxr_name)
names(sxr_name) <- sxr_meta$id
chip_atlas$name <- sxr_name[chip_atlas$name]
rm(ambiguous_srx, ambiguous_tf, sxr_meta_file, sxr_meta, sxr_name)
```

```{r interaction_matrices, echo=FALSE}
gtf <- rtracklayer::readGFF(file = system.file("inst",
                                               "extdata",
                                               params$gtf,
                                               package = "xcore"))
# filter fantomcat by gtf
sel <- gtf %>% 
  dplyr::filter(type == "transcript") %>%
  dplyr::pull(transcript_id)
fantomcat <- fantomcat[transcriptID %in% sel, ]

# create fantomcat GRanges object
fantomcat_gr <-
  GenomicRanges::makeGRangesFromDataFrame(
    df = fantomcat,
    keep.extra.columns = TRUE,
    seqinfo = GenomeInfoDb::Seqinfo(genome = params$genome)
  )

remap_promoters <- getInteractionMatrix(fantomcat_gr, remap)
rm(remap)
gc(verbose = FALSE)

chip_atlas_promoters <- getInteractionMatrix(fantomcat_gr, chip_atlas)
rm(chip_atlas)
gc(verbose = FALSE)
```


```{r report_data, echo=FALSE, message=FALSE}
# Internally used data
# ENCODE blacklist
blacklist <- rtracklayer::import(system.file("inst",
                                             "extdata",
                                             "hg38-blacklist.v2.bed.gz",
                                             package = "xcore")) %>%
    intersectGr(subject = fantomcat_gr, 
                type = "any") %>%
    GenomicRanges::mcols() %>%
    `[[`("name")
    

# ENSEMBL soft mask list 4725854 regions spanning ~50% of human genome
ensembl_sm <- rtracklayer::import(system.file("inst",
                                              "extdata",
                                              "hg38_ensembl_sm.bed.gz",
                                              package = "xcore")) %>%
    intersectGr(subject = fantomcat_gr, 
                type = "any") %>%
    GenomicRanges::mcols() %>%
    `[[`("name")
```
# Technical biases and noise vs biological signal.

## Looking at promoter TF interaction matrix row wise (promoter wise)

`r params$main_string` included `r nrow(remap_promoters)` promoters and
`r ncol(remap_promoters)`, `r ncol(chip_atlas_promoters)` ChIP-seq experiments
for ReMap2020, ChIP-Atlas respectively.

<!-- To narrow down this list we focus on one transcript per gene based on promoter -->
<!-- score. Additionally we look only on genes from ENSEMBL annotation. -->

<!-- ```{r} -->
<!-- best_promoters <- GenomicRanges::mcols(fantomcat_gr) %>% -->
<!--     as.data.frame() %>% -->
<!--     dplyr::group_by(geneID) %>% -->
<!--     dplyr::slice(which.max(score)) %>% -->
<!--     dplyr::filter(grepl(pattern = "^ENSG", x = geneID)) %>% -->
<!--     dplyr::pull(name) -->

<!-- # restrict interaction matrices to core promoters only -->
<!-- remap_promoters <- remap_promoters[best_promoters, ] -->
<!-- chip_atlas_promoters <- chip_atlas_promoters[best_promoters, ] -->
<!-- ``` -->

<!-- This give us `r nrow(remap_promoters)` promoters and -->
<!-- `r ncol(remap_promoters)`, `r ncol(chip_atlas_promoters)` ChIP-seq experiments -->
<!-- for ReMap2020, ChIP-Atlas respectively. -->

### Promoter coverage

For each promoter we calculate coverage which is equal to the number
of ChIP-seq experiment the promoter intersect with. Additionally we calculate 
the TF coverage equal to the number of different TFs the promoter intersect 
with. And biological context coverage which is equal to the number of different 
biological contexts (eg. tissue, cell) promoter intersect with. 

```{r promoters_coverage}
promoters <- list(
    remap = data.table::data.table(
        nm = remap_promoters %>% rownames(),
        coverage = remap_promoters %>% Matrix::rowSums(),
        tf_coverage = remap_promoters %>% getCoverage(col = 2, sep = "\\."),
        con_coverage = remap_promoters %>% getCoverage(col = 3, sep = "\\.")
    ),
    chip_atlas = data.table::data.table(
        nm = chip_atlas_promoters %>% rownames(),
        coverage = chip_atlas_promoters %>% Matrix::rowSums(),
        tf_coverage = chip_atlas_promoters %>%
            getCoverage(col = 1, sep = "_"),
        con_coverage = chip_atlas_promoters %>%
            getCoverage(col = 3, sep = "_")
    )
)
```

```{r promoters_coverage_hist}
# Promoter coverage histogram
par(mfrow = c(1, 2), cex = 0.7, oma = c(0, 0, 2, 0))
coverageHist(x = promoters[["remap"]][["coverage"]],
             main = "ReMap2020 promoter coverage")
coverageHist(x = promoters[["chip_atlas"]][["coverage"]], 
             main = "ChIP-Atlas promoter coverage")
mtext(paste0("FANTOM CAT ", params$main_string), outer = TRUE)

# Promoter TF coverage histogram
coverageHist(x = promoters[["remap"]][["tf_coverage"]],
             main = "ReMap2020 promoter TF coverage")
coverageHist(x = promoters[["chip_atlas"]][["tf_coverage"]], 
             main = "ChIP-Atlas promoter TF coverage")
mtext(paste0("FANTOM CAT ", params$main_string), outer = TRUE)

# Promoter biological context coverage histogram
coverageHist(x = promoters[["remap"]][["con_coverage"]],
             main = "ReMap2020 promoter context coverage")
coverageHist(x = promoters[["chip_atlas"]][["con_coverage"]], 
             main = "ChIP-Atlas promoter context coverage")
mtext(paste0("FANTOM CAT ", params$main_string), outer = TRUE)
```

For some promoters we observe no coverage at all:
`r sum(promoters[["remap"]][["coverage"]] == 0)`,
`r sum(promoters[["chip_atlas"]][["coverage"]] == 0)`
for ReMap2020, ChIP-Atlas respectively.

### ENCODE blacklist

We investigate promoters coverage by overlapping the promoters with ENCODE's 
list of blacklisted regions.

```{r blacklisted_promoters}
par(mfrow = c(1, 2), cex = 0.7, oma = c(0, 0, 2, 0))
blacklistBoxPlot(coverage = promoters[["remap"]][["coverage"]],
                 cov_names = promoters[["remap"]][["nm"]],
                 gr1 = blacklist,
                 aname = "blacklisted",
                 bname = "other",
                 main = "ReMap2020\nENCODE blacklisted regions",
                 ylab = "promoter coverage")
blacklistBoxPlot(coverage = promoters[["chip_atlas"]][["coverage"]],
                 cov_names = promoters[["chip_atlas"]][["nm"]],
                 gr1 = blacklist,
                 aname = "blacklisted",
                 bname = "other",
                 main = "ChIP-Atlas\nENCODE blacklisted regions",
                 ylab = "promoter coverage")
mtext(paste0("FANTOM CAT ", params$main_string), outer = TRUE)
```

### Ensembl soft masked regions

We investigate promoters coverage by overlapping the promoters with Ensembl's 
list of soft masked regions.

```{r soft_masked_promoters}
par(mfrow = c(1, 2), cex = 0.7, oma = c(0, 0, 2, 0))
blacklistBoxPlot(coverage = promoters[["remap"]][["coverage"]],
                 cov_names = promoters[["remap"]][["nm"]],
                 gr1 = ensembl_sm,
                 aname = "soft masked",
                 bname = "other",
                 main = "ReMap2020\nEnsembl soft masked regions",
                 ylab = "promoter coverage")
blacklistBoxPlot(coverage = promoters[["chip_atlas"]][["coverage"]],
                 cov_names = promoters[["chip_atlas"]][["nm"]],
                 gr1 = ensembl_sm,
                 aname = "soft masked",
                 bname = "other",
                 main = "ChIP-Atlas\nEnsembl soft masked regions",
                 ylab = "promoter coverage")
mtext(paste0("FANTOM CAT ", params$main_string), outer = TRUE)
```

### Promoters "phisical" properties

We investigate promoters coverage by looking at the relations between promoter
coverage and promoter FANTOM5 score, or width.

```{r promoter_score_length}
promoter_score <- stats::setNames(log10(fantomcat_gr$score), fantomcat_gr$name)
promoter_width <-
  stats::setNames(GenomicRanges::width(fantomcat_gr), fantomcat_gr$name)

# Promoters coverage vs score
par(mfrow = c(1, 2), cex = 0.7, oma = c(0, 0, 2, 0))
graphics::smoothScatter(
     x = promoters[["remap"]][["coverage"]],
     y = promoter_score[promoters[["remap"]][["nm"]]],
     main = sprintf("ReMap2020\npromoters coverage vs F5 score\nSpearman: %f",
                    cor(x = promoters[["remap"]][["coverage"]],
                        y = promoter_score[promoters[["remap"]][["nm"]]],
                        method = "spearman")),
     xlab = "promoter coverage",
     ylab = "log10(promoter score)")
graphics::smoothScatter(
     x = promoters[["chip_atlas"]][["coverage"]],
     y = promoter_score[promoters[["chip_atlas"]][["nm"]]],
     main = sprintf("ChIP-Atlas\npromoters coverage vs F5 score\nSpearman: %f",
                    cor(x = promoters[["chip_atlas"]][["coverage"]],
                        y = promoter_score[promoters[["chip_atlas"]][["nm"]]],
                        method = "spearman")),
     xlab = "promoter coverage",
     ylab = "log10(promoter score)")
mtext(paste0("FANTOM CAT ", params$main_string), outer = TRUE)

# Promoters coverage vs width
graphics::smoothScatter(
     x = promoters[["remap"]][["coverage"]],
     y = promoter_width[promoters[["chip_atlas"]][["nm"]]],
     main = sprintf("ReMap2020\npromoters coverage vs width\nSpearman: %f",
                    cor(x = promoters[["remap"]][["coverage"]],
                        y = promoter_width[promoters[["chip_atlas"]][["nm"]]],
                        method = "spearman")),
     xlab = "promoter coverage",
     ylab = "promoter width")
graphics::smoothScatter(
     x = promoters[["chip_atlas"]][["coverage"]],
     y = promoter_width[promoters[["chip_atlas"]][["nm"]]],
     main = sprintf("ChIP-Atlas\npromoters coverage vs width\nSpearman: %f",
                    cor(x = promoters[["chip_atlas"]][["coverage"]],
                        y = promoter_width[promoters[["chip_atlas"]][["nm"]]],
                        method = "spearman")),
     xlab = "promoter coverage",
     ylab = "promoter width")
mtext(paste0("FANTOM CAT ", params$main_string), outer = TRUE)
```

### FANTOM5 promoter specificity

```{r promoters_f5_tau_hist}
fantomcat_expression[promoters[["remap"]], 
                     remap_coverage := coverage,
                     on = c("prmtrID" = "nm")]
fantomcat_expression[promoters[["chip_atlas"]], 
                     chip_atlas_coverage := coverage,
                     on = c("prmtrID" = "nm")]
m <- fantomcat_expression$prmtrID %in% promoters[["remap"]][["nm"]]

par(cex = 0.7, oma = c(0, 0, 2, 0))
coverageHist(x = fantomcat_expression[m, tau],
             main = expression(paste("core promoters ", tau, " distribution")))
mtext(paste0("FANTOM CAT ", params$main_string), outer = TRUE)
```

```{r promoter_f5_specificity_scatter}
par(mfrow = c(1, 2), cex = 0.7, oma = c(0, 0, 2, 0))
with(
  fantomcat_expression[m, ],
  graphics::smoothScatter(
    x = remap_coverage,
    y = tau,
    main = sprintf(
      "ReMap2020\npromoter specificity by tau\nSpearman: %f",
      cor(x = remap_coverage,
          y = tau,
          method = "spearman")
    ),
    ylab = expression(tau),
    xlab = "promoter coverage"
  )
)
with(
  fantomcat_expression[m, ],
  graphics::smoothScatter(
    x = chip_atlas_coverage,
    y = tau,
    main = sprintf(
      "ChIP-Atlas\npromoter specificity by tau\nSpearman: %f",
      cor(x = chip_atlas_coverage,
          y = tau,
          method = "spearman")
    ),
    ylab = expression(tau),
    xlab = "promoter coverage"
  )
)
mtext(paste0("FANTOM CAT ", params$main_string), outer = TRUE)
```
