---
title: "Deeper look into the 'signatures'"
author: "Maciej Migda≈Ç"
output: pdf_document
params:
  promoters_ann: promoters_f5
  enhancers_ann: enhancers_f5
  annotation_name: FANTOM5
---

```{r include=FALSE}
knitr::opts_chunk$set(warning = FALSE)

devtools::load_all()
library(data.table)

promoters_ann <- get0(params$promoters_ann)
enhancers_ann <- get0(params$enhancers_ann)

stopifnot(
  all(c(
    "SYMBOL",
    "gene_type_gencode",
    "score",
    "name",
    "tau",
    "protein_atlas_tissue_specificity",
    "encode_blacklist",
    "ensembl_sm"
  ) %in% colnames(GenomicRanges::mcols(promoters_ann)))
)
# stopifnot(
#   all(c(
#     "tau",
#     "encode_blacklist",
#     "ensembl_sm"
#   ) %in% colnames(GenomicRanges::mcols(enhancers_ann)))
# )
```

# Technical biases and noise vs biological signal.

Based on the Bogumil's observation that some promoters always give a signal
regardless on the transcription factor we look at. As well as, some TF gives 
signal regardless of the promoter we look at. We set out with an exploratory 
data analysis on the ReMap2020 and ChIP-Atlas data sets.

## Looking at promoter TF interaction matrix row wise (promoter wise)

ReMap2020 and ChIP-Atlas were intersected with FANTOM5 promoters into 
interaction matrices with ChIP-seq experiment as a columns and promoters as a
row names. Overall our data included `r nrow(remap_promoters)` promoters and
`r ncol(remap_promoters)`, `r ncol(chip_atlas_promoters)` ChIP-seq experiments
for ReMap2020, ChIP-Atlas respectively.

To look deeper into those data sets, first we have limited our analysis to 
core promoters defined as promoters associated with a protein coding gene (as
per GENCODE 38) and supported by ENCODE ROADMAP data. Finally, we select one
promoter per associated gene that have maximum FANTOM5 score.
   
```{r core_promoters}
# only protein coding genes as per GENCODE annotation
promoters_ann_core <- 
    promoters_ann[promoters_ann$gene_type_gencode == "protein_coding", ]

# ENCODE ROADMAP confirmation
roadmap_promoters <- rtracklayer::import.bed(
    con = system.file(
        "inst",
        "extdata",
        "Epigenome5DRoadmapDHS_promoter_hg38_liftOver.bed",
        package = "xcore"
    )
)
GenomeInfoDb::seqlevels(roadmap_promoters, pruning.mode = "coarse") <-
    GenomeInfoDb::seqlevels(promoters_ann_core)
promoters_ann_core <- intersectGR(promoters_ann_core, roadmap_promoters)

# Select best promoter per gene
best_promoters <- GenomicRanges::mcols(promoters_ann_core) %>%
    as.data.frame() %>%
    dplyr::group_by(SYMBOL) %>%
    dplyr::slice(which.max(score)) %>%
    dplyr::pull(name)
promoters_ann_core <-
    promoters_ann_core[promoters_ann_core$name %in% best_promoters, ]

# restrict interaction matrices to core promoters only
remap_promoters_core <- remap_promoters[promoters_ann_core$name, ]
chip_atlas_promoters_core <- chip_atlas_promoters[promoters_ann_core$name, ]
```

This give us an interaction matrix with `r nrow(remap_promoters_core)` promoters
(rows) and `r ncol(remap_promoters_core)`, `r ncol(chip_atlas_promoters_core)` 
ChIP-seq experiments (columns) for ReMap2020, ChIP-Atlas respectively.

### Promoter coverage

For each promoter we calculate coverage which is equal to the number
of ChIP-seq experiment the promoter intersect with. Additionally we calculate 
the TF coverage equal to the number of different TFs the promoter intersect 
with. And biological context coverage which is equal to the number of different 
biological contexts (eg. tissue, cell) promoter intersect with. 

```{r promoters_coverage}
# remap
promoters_ann_core$remap_coverage <- remap_promoters_core %>% Matrix::rowSums()
# chip_atlas
promoters_ann_core$chip_atlas_coverage <- chip_atlas_promoters_core %>% Matrix::rowSums()

# splitting coverage into tf, biotype or study gives values that are highly 
# correlated with coverage (>0.95) - making it not very interesting to look at.
# TODO revisit after TF filtration.
# cor(promoters_ann_core$remap_coverage, getCoverage(
#           mat = remap_promoters_core, 
#           gr = remap_meta[colnames(remap_promoters_core), tf, on = "id"]))
# cor(promoters_ann_core$remap_coverage, getCoverage(
#           mat = remap_promoters_core, 
#           gr = remap_meta[colnames(remap_promoters_core), biotype, on = "id"]))
# cor(promoters_ann_core$remap_coverage, getCoverage(
#           mat = remap_promoters_core, 
#           gr = remap_meta[colnames(remap_promoters_core), study, on = "id"]))
```

```{r promoters_coverage_hist}
par(mfrow = c(1, 2), cex = 0.7, pty = "s")

# Promoter coverage histogram
plotHistogram(x = promoters_ann_core$remap_coverage,
             main = "ReMap2020 promoter coverage")
plotHistogram(x = promoters_ann_core$chip_atlas_coverage, 
             main = "ChIP-Atlas promoter coverage")
```

For some promoters we observe no coverage at all:
`r sum(promoters[["remap"]][["coverage"]] == 0)`,
`r sum(promoters[["chip_atlas"]][["coverage"]] == 0)`
for ReMap2020, ChIP-Atlas respectively.


### Protein Atlas gene specificity classification

We investigate promoters coverage by looking at their target genes 
classification in terms of tissue/cell type specificity as defined by Protein 
Atlas.

> The RNA specificity category is based on mRNA expression levels in a 
combination of data from HPA, GTEX and FANTOM5. The categories include: tissue enriched, group enriched, tissue enhanced, low tissue specificity and not 
detected.

>
+ Enriched - NX level in a particular tissue/region/cell type at least four 
times any other tissue/cell type.
+ Group enriched - NX levels of a group (2-5 tissues or 2-10 cell types) at 
least four times any other tissue/cell type.
+ Enhanced - NX levels of a group (1-5 tissues or 1-10 cell types) at least four 
times the mean of other tissue/cell types.
+ Low specificity - NX >= 1 in at least one tissue/cell type but not elevated in 
any tissue/cell type.
+ Not detected - NX < 1 in all tissue/cell types.

[Protein Atlas](https://www.proteinatlas.org/about/assays+annotation#classification_rna)

```{r protein_atlas_promoters}
# RNA tissue specificity
par(mfrow = c(1, 2), cex = 0.7, pty = "s")
ymax <-
  max(promoters_ann_core$remap_coverage,
      promoters_ann_core$chip_atlas_coverage)
graphics::boxplot(remap_coverage ~ protein_atlas_cell_specificity, 
                  data = promoters_ann_core,
                  ylim = c(0, ymax),
                  main = "ReMap2020",
                  ylab = "promoter coverage",
                  xlab = "",
                  las=2,
                  cex.names = 0.4)
graphics::boxplot(chip_atlas_coverage ~ protein_atlas_cell_specificity, 
                  data = promoters_ann_core,
                  ylim = c(0, ymax),
                  main = "ChIP-Atlas",
                  ylab = "promoter coverage",
                  xlab = "",
                  las=2,
                  cex.names = 0.4)
mtext(
  "Protein Atlas tissue specificity",
  side = 3,
  outer = TRUE,
  line = -2,
  font = 2,
  cex = 1.2
)
```

### Promoter tissue specificity

To generalize the idea of promoter tissue/cell type specificity to promoters 
outside Protein Atlas annotation we define specificity score based on the 
FANTOM5 CAGE data.

The  can be defined in multiple ways. Here we calculate expression based 
specificity score using $\tau$ metric defined as

$\tau = \frac{\sum_{i=1}^{n} (1-\hat{x_i})}{n-1}; \hat{x_i} = \frac{x_i}{max(x)}$

Tau varies from 0 to 1, where 0 means broadly expressed, and 1 is specific.

[reference](https://academic.oup.com/bib/article/18/2/205/2562739#119555186).

```{r promoter_f5_specificity_boxplot}
graphics::boxplot(tau ~ protein_atlas_tissue_specificity, 
                  data = promoters_ann_core,
                  main = "Tau vs Protein Atlass classififcation",
                  ylab = expression(tau),
                  xlab = "",
                  cex.axis = 0.8)
```


```{r promoters_f5_tau_hist}
plotHistogram(x = promoters_ann_core$tau,
             main = expression(paste("core promoters ", tau, " distribution")))
```

```{r promoter_f5_specificity_scatter}
par(mfrow = c(1, 2), cex = 0.7, pty = "s", mgp=c(2,1,0))
with(
  as.data.frame(promoters_ann_core),
  graphics::smoothScatter(
    x = remap_coverage,
    y = tau,
    main = sprintf(
      "ReMap2020\nSpearman: %f",
      cor(x = remap_coverage,
          y = tau,
          method = "spearman")
    ),
    ylab = expression(tau),
    xlab = "promoter coverage"
  )
)
with(
  as.data.frame(promoters_ann_core),
  graphics::smoothScatter(
    x = chip_atlas_coverage,
    y = tau,
    main = sprintf(
      "ChIP-Atlas\nSpearman: %f",
      cor(x = chip_atlas_coverage,
          y = tau,
          method = "spearman")
    ),
    ylab = expression(tau),
    xlab = "promoter coverage"
  )
)
mtext(
  "Promoter coverage vs tau",
  side = 3,
  outer = TRUE,
  line = -2,
  font = 2,
  cex = 1.2
)
```

### ENCODE blacklist

We investigate promoters coverage by overlapping the promoters with ENCODE's 
list of blacklisted regions.

>  ENCODE blacklist is a comprehensive set of regions in the human and other 
genomes that have anomalous, unstructured, or high signal in next-generation 
sequencing experiments independent of cell line or experiment. The removal of 
the ENCODE blacklist is an essential quality measure when analyzing functional 
genomics data.
[ENCODE blacklist](https://github.com/Boyle-Lab/Blacklist)

```{r blacklisted_promoters}
par(mfrow = c(1, 2), cex = 0.7, pty = "s")
ymax <-
  max(promoters_ann_core$remap_coverage,
      promoters_ann_core$chip_atlas_coverage)
with(
  as.data.frame(promoters_ann_core),
  boxplot(
    remap_coverage[encode_blacklist],
    remap_coverage[!encode_blacklist],
    names = c(sprintf(
      "blacklisted (%i)", sum(encode_blacklist)
    ),
    sprintf("other (%i)", sum(!encode_blacklist))),
    ylim = c(0, ymax),
    main = "ReMap2020",
    ylab = "promoter coverage"
  )
)
with(
  as.data.frame(promoters_ann_core),
  boxplot(
    chip_atlas_coverage[encode_blacklist],
    chip_atlas_coverage[!encode_blacklist],
    names = c(sprintf(
      "blacklisted (%i)", sum(encode_blacklist)
    ),
    sprintf("other (%i)", sum(!encode_blacklist))),
    ylim = c(0, ymax),
    main = "ChIP-Atlas",
    ylab = "promoter coverage"
  )
)
mtext(
  "ENCODE blacklisted regions",
  side = 3,
  outer = TRUE,
  line = -2,
  font = 2,
  cex = 1.2
)
```

### Promoters score (width) vs coverage

We investigate promoters coverage by looking at the relations between promoter
coverage and score, or width.

```{r promoter_score_length}
# Promoters coverage vs score
par(mfrow = c(1, 2), cex = 0.7, pty = "s")
graphics::smoothScatter(
     x = promoters_ann_core$remap_coverage,
     y = log10(promoters_ann_core$score),
     main = sprintf("ReMap2020\nSpearman: %f",
                    cor(x = promoters_ann_core$remap_coverage,
                        y = log10(promoters_ann_core$score),
                        method = "spearman")),
     xlab = "promoter coverage",
     ylab = "log10(promoter score)")
graphics::smoothScatter(
     x = promoters_ann_core$chip_atlas_coverage,
     y = log10(promoters_ann_core$score),
     main = sprintf("ChIP-Atlas\nSpearman: %f",
                    cor(x = promoters_ann_core$chip_atlas_coverage,
                        y = log10(promoters_ann_core$score),
                        method = "spearman")),
     xlab = "promoter coverage",
     ylab = "log10(promoter score)")
mtext(
  "Promoter coverage vs promoter score",
  side = 3,
  outer = TRUE,
  line = -2,
  font = 2,
  cex = 1.2
)

# Promoters coverage vs width
graphics::smoothScatter(
     x = promoters_ann_core$remap_coverage,
     y = GenomicRanges::width(promoters_ann_core),
     main = sprintf("ReMap2020\nSpearman: %f",
                    cor(x = promoters_ann_core$remap_coverage,
                        y = GenomicRanges::width(promoters_ann_core),
                        method = "spearman")),
     xlab = "promoter coverage",
     ylab = "promoter width")
graphics::smoothScatter(
     x = promoters_ann_core$chip_atlas_coverage,
     y = GenomicRanges::width(promoters_ann_core),
     main = sprintf("ChIP-Atlas\nSpearman: %f",
                    cor(x = promoters_ann_core$chip_atlas_coverage,
                        y = GenomicRanges::width(promoters_ann_core),
                        method = "spearman")),
     xlab = "promoter coverage",
     ylab = "promoter width")
mtext(
  "Promoter coverage vs promoter width",
  side = 3,
  outer = TRUE,
  line = -2,
  font = 2,
  cex = 1.2
)
```

## Looking at promoter TF interaction matrix column wise (TF wise)

### TF overlapp between ReMap2020 and ChIP-Atlas

```{r}
# ReMap2020
table((remap_meta$tf %in% chip_atlas_meta$tf)) %>%
  `/`(length(remap_meta$tf)) %>% 
  data.frame() %>% 
  dplyr::rename(TF = Var1) %>% 
  dplyr::mutate(TF = c("not included in ChIP-Atlas", "included in ChIP-Atlas")) %>%
  knitr::kable(caption = "ReMap2020 TF")

table(unique(remap_meta$tf) %in% unique(chip_atlas_meta$tf)) %>%
  `/`(length(unique(remap_meta$tf))) %>% 
  data.frame() %>% 
  dplyr::rename(TF = Var1) %>% 
  dplyr::mutate(TF = c("not included in ChIP-Atlas", "included in ChIP-Atlas")) %>%
  knitr::kable(caption = "ReMap2020 unique TF")

remap_meta$tf[! remap_meta$tf %in% chip_atlas_meta$tf] %>%
  table() %>%
  as.numeric() %>%
  plotHistogram(breaks = "Sturges", 
               main = "ReMap2020 TF not included in ChIP-Atlas")

# ChIP-Atlas
table((chip_atlas_meta$tf %in% remap_meta$tf)) %>%
  `/`(length(chip_atlas_meta$tf)) %>% 
  data.frame() %>% 
  dplyr::rename(TF = Var1) %>% 
  dplyr::mutate(TF = c("not included in ReMap2020", "included in ReMap2020")) %>%
  knitr::kable(caption = "ChIP-Atlas TF")

table(unique(chip_atlas_meta$tf) %in% unique(remap_meta$tf)) %>%
  `/`(length(unique(chip_atlas_meta$tf))) %>% 
  data.frame() %>% 
  dplyr::rename(TF = Var1) %>% 
  dplyr::mutate(TF = c("not included in ChIP-Atlas", "included in ChIP-Atlas")) %>%
  knitr::kable(caption = "ChIP-Atlas unique TF")

chip_atlas_meta$tf[! chip_atlas_meta$tf %in% remap_meta$tf] %>%
  table() %>%
  as.numeric() %>%
  plotHistogram(breaks = "Sturges", 
               main = "ChIP-Atlas TF not included in ChIP-Atlas")
```

```{r}
remap_meta[, coverage := Matrix::colSums(remap_promoters_core[, id])]
chip_atlas_meta[, coverage := Matrix::colSums(chip_atlas_promoters_core[, id])]
```

For the basic analysis we have focused only on the ReMap dataset. Choosing only
experiments that overlap more than 1% of core promoters and without any 
condition or treatment label.

Further we focused only on TF with >= 10 experiments and biotypes with >= 29 
experiments, excluding the "breast" and "liver" biotypes.

```{r}
remap_meta_core <- remap_meta[(coverage >= nrow(remap_promoters_core) * 0.01) &
                                (condition == "")]

sel_tf <- remap_meta_core$tf %>% 
  table() %>% 
  `[`(. >= 10) %>% 
  names()
remap_meta_core <- remap_meta_core[tf %in% sel_tf]

sel_biotypes <- remap_meta_core$biotype %>% 
  table() %>% 
  `[`(. >= 29) %>% 
  names() %>%
  `[`(! . %in% c("breast", "liver"))
remap_meta_core <- remap_meta_core[biotype %in% sel_biotypes]

# annotate coverage
remap_meta_core[, cov_type := cut(
  coverage,
  breaks = c(0, quantile(coverage, 0.25), quantile(coverage, 0.75), Inf),
  labels = c("low", "medium", "high")
)]
```

```{r}
plotHistogram(remap_meta_core$coverage, 
              main = "ReMap2020 selected experiment's coverage")
```

### TF / biotype confusion matrix

```{r tf_bg_confusion_matrix, echo=FALSE}
remap_tfs_tab <-
  data.table::dcast(data = remap_meta_core,
                    formula = biotype ~ tf,
                    fun = length)
ord <- c(1, order(remap_tfs_tab[, -1] %>% colSums(), decreasing = TRUE) + 1)
data.table::setcolorder(remap_tfs_tab, ord)
remap_tfs_tab[, .r := -rowSums(remap_tfs_tab[, -1])]
data.table::setorder(remap_tfs_tab, .r)[, .r := NULL]

knitr::kable(remap_tfs_tab[, 1:21], caption = "TF/biotype confusion matrix, top 20 TF")
```

### MYC

```{r}
breaks <- seq(from = -1, to = 1, by = 0.1)
color <-
  grDevices::colorRampPalette(rev(RColorBrewer::brewer.pal(
    n = 7, 
    name = "RdYlBu"))
  )(length(breaks))
  
plotExperimentsHeatmap(
  mat = remap_promoters_core,
  meta = remap_meta_core[tf == "MYC", -c("tf", "condition", "tf_dbd")],
  breaks = breaks,
  color = color,
  labels_row = remap_meta_core[tf == "MYC", biotype],
  labels_col = remap_meta_core[tf == "MYC", biotype],
  main = "MYC experiments Pearson correlation")
```

### MAX

```{r}
breaks <- seq(from = -1, to = 1, by = 0.1)
color <-
  grDevices::colorRampPalette(rev(RColorBrewer::brewer.pal(
    n = 7, 
    name = "RdYlBu"))
  )(length(breaks))
  
plotExperimentsHeatmap(
  mat = remap_promoters_core,
  meta = remap_meta_core[tf == "MAX", -c("tf", "condition", "tf_dbd")],
  breaks = breaks,
  color = color,
  labels_row = remap_meta_core[tf == "MAX", biotype],
  labels_col = remap_meta_core[tf == "MAX", biotype],
  main = "MAX experiments Pearson correlation")
```

### ELF1

```{r}
breaks <- seq(from = -1, to = 1, by = 0.1)
color <-
  grDevices::colorRampPalette(rev(RColorBrewer::brewer.pal(
    n = 7, 
    name = "RdYlBu"))
  )(length(breaks))

plotExperimentsHeatmap(
  mat = remap_promoters_core,
  meta = remap_meta_core[tf == "ELF1", -c("tf", "condition", "tf_dbd")],
  breaks = breaks,
  color = color,
  labels_row = remap_meta_core[tf == "ELF1", biotype],
  labels_col = remap_meta_core[tf == "ELF1", biotype],
  main = "ELF1 experiments Pearson correlation")
```

### FOXA1

```{r}
breaks <- seq(from = -1, to = 1, by = 0.1)
color <-
  grDevices::colorRampPalette(rev(RColorBrewer::brewer.pal(
    n = 7, 
    name = "RdYlBu"))
  )(length(breaks))

plotExperimentsHeatmap(
  mat = remap_promoters_core,
  meta = remap_meta_core[tf == "FOXA1", -c("tf", "condition", "tf_dbd")],
  breaks = breaks,
  color = color,
  labels_row = remap_meta_core[tf == "FOXA1", biotype],
  labels_col = remap_meta_core[tf == "FOXA1", biotype],
  main = "FOXA1 experiments Pearson correlation")
```

### Core experiments heatmap

```{r include=FALSE}
breaks <- seq(from = -1, to = 1, by = 0.1)
color <-
  grDevices::colorRampPalette(rev(RColorBrewer::brewer.pal(
    n = 7, 
    name = "RdYlBu"))
  )(length(breaks))

plotExperimentsHeatmap(
  mat = remap_promoters_core,
  meta = remap_meta_core[, -c("study", "condition")],
  breaks = breaks,
  color = color,
  labels_row = remap_meta_core[, paste(tf, biotype)],
  labels_col = remap_meta_core[, paste(tf, biotype)],
  main = "Core experiments Pearson correlation")
```

## Clustering experiments

```{r}
mat <- remap_promoters_core[, remap_meta_core[, id]]
mat_cor <- cor(mat %>% as.matrix(), method = "pearson")
d <- as.dist(abs((1 - mat_cor) / 2))
hc.complete <- hclust(d, method = "complete")

plot(
  hc.complete,
  labels = FALSE,
  main = "Complete Linkage ",
  xlab = "",
  sub = "",
  cex = .9)

measureClust <- function(c1, hc, N) {
  nclust_measure <- lapply(
    X = N,
    FUN = function(n) {
      do.call(cbind,
              aricode::clustComp(c1 = c1,
                                 c2 = cutree(hc, n)))
    }
  )
  do.call(rbind, nclust_measure)
}

nclust <- seq(from = 1, to = 400, by = 5)
nclust_measure_tf <-
  measureClust(remap_meta_core[, tf], hc.complete, nclust)
nclust_measure_biotype <-
  measureClust(remap_meta_core[, biotype], hc.complete, nclust)
nclust_measure_tf_biotype <-
  measureClust(remap_meta_core[, paste0(tf, biotype)], hc.complete, nclust)
nclust_measure_study <-
  measureClust(remap_meta_core[, study], hc.complete, nclust)
nclust_measure_dbd <-
  measureClust(remap_meta_core[, tf_dbd], hc.complete, nclust)

# Adjusted Rand Index (ARI)
# Low value indicates that the two data clustering do not agree on any pair of
# points and 1 indicates that the data clustering are exactly the same.
measure <- "ARI"
plot(nclust, nclust_measure_tf[, measure], col = "red", 
     ylim = c(0, 0.5), xlab = "number of clusters", ylab = measure)
points(nclust, nclust_measure_biotype[, measure], col = "blue")
points(nclust, nclust_measure_tf_biotype[, measure], col = "green")
points(nclust, nclust_measure_study[, measure], col = "black")
points(nclust, nclust_measure_dbd[, measure], col = "orange")
legend(
  "topleft",
  legend = c("TF", "biotype", "TF-biotype", "study", "DBD"),
  fill = c("red", "blue", "green", "black", "orange")
)

# Adjusted Mutual Information (AMI)
# measure of clustering quality 0 no mutual information and 1 perfect
# correlation; adjusted for chance
measure <- "NMI"
plot(nclust, nclust_measure_tf[, measure], col = "red", 
     ylim = c(0, 1), xlab = "number of clusters", ylab = measure)
points(nclust, nclust_measure_biotype[, measure], col = "blue")
points(nclust, nclust_measure_tf_biotype[, measure], col = "green")
points(nclust, nclust_measure_study[, measure], col = "black")
points(nclust, nclust_measure_dbd[, measure], col = "orange")
legend(
  "topleft",
  legend = c("TF", "biotype", "TF-biotype", "study", "DBD"),
  fill = c("red", "blue", "green", "black", "orange")
)

hc.clusters <- cutree(hc.complete, 106)
table(hc.clusters, remap_meta_core[, tf])
```


### Experiments correlations in bulk

```{r explore_correlation}
mat <- remap_promoters_core[, remap_meta_core[, id]]
mat_cor <- cor(mat %>% as.matrix(), method = "pearson")

getCors <- function(mat, patterns, invert=FALSE) {
  sapply(
    X = patterns, 
    function(x) {
      m_row <- grep(x, rownames(mat), value = TRUE)
      m_col <- grep(x, colnames(mat), value = TRUE, invert = invert)
      sel_cor_mat <- mat[m_row, m_col]
      sel_cor_mat[upper.tri(sel_cor_mat, diag = FALSE)]
    })
}

# tf
same_tf_cor <- getCors(mat_cor, remap_meta_core[, unique(tf)])
diff_tf_cor <- getCors(mat_cor, remap_meta_core[, unique(tf)], invert = TRUE)
boxplot(unlist(same_tf_cor), unlist(diff_tf_cor), 
        names = c("same TF", "different TFs"), 
        main = "Correlation between experiments")

# biotype
same_bio_cor <- getCors(mat_cor, remap_meta_core[, unique(biotype)])
diff_bio_cor <- getCors(mat_cor, remap_meta_core[, unique(biotype)], invert = TRUE)
boxplot(unlist(same_bio_cor), unlist(diff_bio_cor), 
        names = c("same biotype", "different biotype"), 
        main = "Correlation between experiments")

# tf-biotype
same_tfbio_cor <- getCors(mat_cor, remap_meta_core[, unique(paste0(tf, ".", biotype))])
diff_tfbio_cor <- getCors(mat_cor, remap_meta_core[, unique(paste0(tf, ".", biotype))], invert = TRUE)
boxplot(unlist(same_tfbio_cor), unlist(diff_tfbio_cor), 
        names = c("same TF-biotype", "different TF-biotype"), 
        main = "Correlation between experiments")

# same study
same_study_cor <- getCors(mat_cor, remap_meta_core[, unique(study)])
diff_sutdy_cor <- getCors(mat_cor, remap_meta_core[, unique(study)], invert = TRUE)
boxplot(unlist(same_study_cor), unlist(diff_sutdy_cor), 
        names = c("same study", "different study"), 
        main = "Correlation between experiments")
```

## Experiment collapsing

```{r include=FALSE}
core_id <- colnames(remap_promoters)

remap_promoters_core_tf <- simplifyInteractionMatrix(
  mat = remap_promoters_core,
  gr = remap_meta[id %in% core_id, tf]
)

remap_meta[, tf_biotype := paste0(tf, "_", biotype)]
sel_bio <- remap_meta[id %in% core_id, .N, by = tf_biotype][N >= 3, tf_biotype]
remap_promoters_core_tf_biotype <- simplifyInteractionMatrix(
  mat = remap_promoters_core[, 
             remap_meta[(id %in% core_id) & (tf_biotype %in% sel_bio), id]],
  gr = remap_meta[(id %in% core_id) & (tf_biotype %in% sel_bio), tf_biotype])

promoters_ann_core$remap_tf_coverage <- Matrix::rowSums(remap_promoters_core_tf)
promoters_ann_core$remap_tf_biotype_coverage <-
  Matrix::rowSums(remap_promoters_core_tf_biotype)
with(
  promoters_ann_core,
  graphics::smoothScatter(
    x = remap_promoters_core_tf,
    y = tau,
    main = sprintf(
      "ReMap2020\nSpearman: %f",
      cor(x = remap_promoters_core_tf,
          y = tau,
          method = "spearman")
    ),
    ylab = expression(tau),
    xlab = "promoter coverage"
  )
)
```

