---
title: "Deeper look into the 'signatures'"
author: "Maciej Migda≈Ç"
output: pdf_document
params:
  promoters_ann: promoters_f5
  enhancers_ann: enhancers_f5
  annotation_name: FANTOM5
---

```{r include=FALSE}
knitr::opts_chunk$set(warning = FALSE)

devtools::load_all()
library(data.table)

promoters_ann <- get0(params$promoters_ann)
enhancers_ann <- get0(params$enhancers_ann)

stopifnot(
  all(c(
    "SYMBOL",
    "gene_type_gencode",
    "score",
    "name",
    "tau",
    "protein_atlas_tissue_specificity",
    "encode_blacklist",
    "ensembl_sm"
  ) %in% colnames(GenomicRanges::mcols(promoters_ann)))
)
# stopifnot(
#   all(c(
#     "tau",
#     "encode_blacklist",
#     "ensembl_sm"
#   ) %in% colnames(GenomicRanges::mcols(enhancers_ann)))
# )
```

# Technical biases and noise vs biological signal.

Based on the Bogumil's observation that some promoters always give a signal
regardless on the transcription factor we look at. As well as, some TF gives 
signal regardless of the promoter we look at. We set out with an exploratory 
data analysis on the ReMap2020 and ChIP-Atlas data sets.

## Looking at promoter TF interaction matrix row wise (promoter wise)

ReMap2020 and ChIP-Atlas were intersected with FANTOM5 promoters into 
interaction matrices with ChIP-seq experiment as a columns and promoters as a
row names. Overall our data included `r nrow(remap_promoters)` promoters and
`r ncol(remap_promoters)`, `r ncol(chip_atlas_promoters)` ChIP-seq experiments
for ReMap2020, ChIP-Atlas respectively.

To look deeper into those data sets, first we have limited our analysis to 
core promoters defined as promoters associated with a protein coding gene (as
per GENCODE 38) and supported by ENCODE ROADMAP data. Finally, we select one
promoter per associated gene that have maximum FANTOM5 score.
   
```{r core_promoters}
# only protein coding genes as per GENCODE annotation
promoters_ann_core <- 
    promoters_ann[promoters_ann$gene_type_gencode == "protein_coding", ]

# ENCODE ROADMAP confirmation
roadmap_promoters <- rtracklayer::import.bed(
    con = system.file(
        "inst",
        "extdata",
        "Epigenome5DRoadmapDHS_promoter_hg38_liftOver.bed",
        package = "xcore"
    )
)
GenomeInfoDb::seqlevels(roadmap_promoters, pruning.mode = "coarse") <-
    GenomeInfoDb::seqlevels(promoters_ann_core)
promoters_ann_core <- intersectGR(promoters_ann_core, roadmap_promoters)

# Select best promoter per gene
best_promoters <- GenomicRanges::mcols(promoters_ann_core) %>%
    as.data.frame() %>%
    dplyr::group_by(SYMBOL) %>%
    dplyr::slice(which.max(score)) %>%
    dplyr::pull(name)
promoters_ann_core <-
    promoters_ann_core[promoters_ann_core$name %in% best_promoters, ]

# restrict interaction matrices to core promoters only
remap_promoters_core <- remap_promoters[promoters_ann_core$name, ]
chip_atlas_promoters_core <- chip_atlas_promoters[promoters_ann_core$name, ]
```

This give us an interaction matrix with `r nrow(remap_promoters_core)` promoters
(rows) and `r ncol(remap_promoters_core)`, `r ncol(chip_atlas_promoters_core)` 
ChIP-seq experiments (columns) for ReMap2020, ChIP-Atlas respectively.

### Promoter coverage

For each promoter we calculate coverage which is equal to the number
of ChIP-seq experiment the promoter intersect with. Additionally we calculate 
the TF coverage equal to the number of different TFs the promoter intersect 
with. And biological context coverage which is equal to the number of different 
biological contexts (eg. tissue, cell) promoter intersect with. 

```{r promoters_coverage}
# remap
promoters_ann_core$remap_coverage <- remap_promoters_core %>% Matrix::rowSums()
# chip_atlas
promoters_ann_core$chip_atlas_coverage <- chip_atlas_promoters_core %>% Matrix::rowSums()

# splitting coverage into tf, biotype or study gives values that are highly 
# correlated with coverage (>0.95) - making it not very interesting to look at.
# TODO revisit after TF filtration.
# cor(promoters_ann_core$remap_coverage, getCoverage(
#           mat = remap_promoters_core, 
#           gr = remap_meta[colnames(remap_promoters_core), tf, on = "id"]))
# cor(promoters_ann_core$remap_coverage, getCoverage(
#           mat = remap_promoters_core, 
#           gr = remap_meta[colnames(remap_promoters_core), biotype, on = "id"]))
# cor(promoters_ann_core$remap_coverage, getCoverage(
#           mat = remap_promoters_core, 
#           gr = remap_meta[colnames(remap_promoters_core), study, on = "id"]))
```

```{r promoters_coverage_hist}
par(mfrow = c(1, 2), cex = 0.7, pty = "s")

# Promoter coverage histogram
plotHistogram(x = promoters_ann_core$remap_coverage,
             main = "ReMap2020 promoter coverage")
plotHistogram(x = promoters_ann_core$chip_atlas_coverage, 
             main = "ChIP-Atlas promoter coverage")
```

For some promoters we observe no coverage at all:
`r sum(promoters[["remap"]][["coverage"]] == 0)`,
`r sum(promoters[["chip_atlas"]][["coverage"]] == 0)`
for ReMap2020, ChIP-Atlas respectively.


### Protein Atlas gene specificity classification

We investigate promoters coverage by looking at their target genes 
classification in terms of tissue/cell type specificity as defined by Protein 
Atlas.

> The RNA specificity category is based on mRNA expression levels in a 
combination of data from HPA, GTEX and FANTOM5. The categories include: tissue enriched, group enriched, tissue enhanced, low tissue specificity and not 
detected.

>
+ Enriched - NX level in a particular tissue/region/cell type at least four 
times any other tissue/cell type.
+ Group enriched - NX levels of a group (2-5 tissues or 2-10 cell types) at 
least four times any other tissue/cell type.
+ Enhanced - NX levels of a group (1-5 tissues or 1-10 cell types) at least four 
times the mean of other tissue/cell types.
+ Low specificity - NX >= 1 in at least one tissue/cell type but not elevated in 
any tissue/cell type.
+ Not detected - NX < 1 in all tissue/cell types.

[Protein Atlas](https://www.proteinatlas.org/about/assays+annotation#classification_rna)

```{r protein_atlas_promoters}
# RNA tissue specificity
par(mfrow = c(1, 2), cex = 0.7, pty = "s")
graphics::boxplot(remap_coverage ~ protein_atlas_cell_specificity, 
                  data = promoters_ann_core,
                  main = "ReMap2020",
                  ylab = "promoter coverage",
                  xlab = "",
                  las=2,
                  cex.names = 0.4)
graphics::boxplot(chip_atlas_coverage ~ protein_atlas_cell_specificity, 
                  data = promoters_ann_core,
                  main = "ChIP-Atlas",
                  ylab = "promoter coverage",
                  xlab = "",
                  las=2,
                  cex.names = 0.4)
mtext(
  "Protein Atlas tissue specificity",
  side = 3,
  outer = TRUE,
  line = -2,
  font = 2,
  cex = 1.2
)
```

### FANTOM5 promoter specificity

To generalize the idea of promoter tissue/cell type specificity to promoters 
outside Protein Atlas annotation we define specificity score based on the 
FANTOM5 CAGE data.

The  can be defined in multiple ways. Here we calculate expression based 
specificity score using $\tau$ metric defined as

$\tau = \frac{\sum_{i=1}^{n} (1-\hat{x_i})}{n-1}; \hat{x_i} = \frac{x_i}{max(x)}$

Tau varies from 0 to 1, where 0 means broadly expressed, and 1 is specific.

[reference](https://academic.oup.com/bib/article/18/2/205/2562739#119555186).

```{r promoter_f5_specificity_boxplot}
graphics::boxplot(tau ~ protein_atlas_tissue_specificity, 
                  data = promoters_ann_core,
                  main = "Tau vs Protein Atlass classififcation",
                  ylab = expression(tau),
                  xlab = "",
                  cex.axis = 0.8)
```


```{r promoters_f5_tau_hist}
plotHistogram(x = promoters_ann_core$tau,
             main = expression(paste("core promoters ", tau, " distribution")))
```

```{r promoter_f5_specificity_scatter}
par(mfrow = c(1, 2), cex = 0.7, pty = "s", mgp=c(2,1,0))
with(
  promoters_ann_core,
  graphics::smoothScatter(
    x = remap_coverage,
    y = tau,
    main = sprintf(
      "ReMap2020\nSpearman: %f",
      cor(x = remap_coverage,
          y = tau,
          method = "spearman")
    ),
    ylab = expression(tau),
    xlab = "promoter coverage"
  )
)
with(
  promoters_ann_core,
  graphics::smoothScatter(
    x = chip_atlas_coverage,
    y = tau,
    main = sprintf(
      "ChIP-Atlas\nSpearman: %f",
      cor(x = chip_atlas_coverage,
          y = tau,
          method = "spearman")
    ),
    ylab = expression(tau),
    xlab = "promoter coverage"
  )
)
mtext(
  "Promoter coverage vs tau",
  side = 3,
  outer = TRUE,
  line = -2,
  font = 2,
  cex = 1.2
)
```

### ENCODE blacklist

We investigate promoters coverage by overlapping the promoters with ENCODE's 
list of blacklisted regions.

>  ENCODE blacklist is a comprehensive set of regions in the human and other 
genomes that have anomalous, unstructured, or high signal in next-generation 
sequencing experiments independent of cell line or experiment. The removal of 
the ENCODE blacklist is an essential quality measure when analyzing functional 
genomics data.
[ENCODE blacklist](https://github.com/Boyle-Lab/Blacklist)

```{r blacklisted_promoters}
par(mfrow = c(1, 2), cex = 0.7, pty = "s")
with(promoters_ann_core, boxplot(
  remap_coverage[encode_blacklist], remap_coverage[! encode_blacklist],
  names = c(sprintf("blacklisted (%i)", sum(encode_blacklist)),
            sprintf("other (%i)", sum(! encode_blacklist))),
  main = "ReMap2020",
  ylab = "promoter coverage"
))
with(promoters_ann_core, boxplot(
  chip_atlas_coverage[encode_blacklist], chip_atlas_coverage[! encode_blacklist],
  names = c(sprintf("blacklisted (%i)", sum(encode_blacklist)),
            sprintf("other (%i)", sum(! encode_blacklist))),
  main = "ChIP-Atlas",
  ylab = "promoter coverage"
))
mtext(
  "ENCODE blacklisted regions",
  side = 3,
  outer = TRUE,
  line = -2,
  font = 2,
  cex = 1.2
)
```

### Promoters score (width) vs coverage

We investigate promoters coverage by looking at the relations between promoter
coverage and score, or width.

```{r promoter_score_length}
# Promoters coverage vs score
par(mfrow = c(1, 2), cex = 0.7, pty = "s")
graphics::smoothScatter(
     x = promoters_ann_core$remap_coverage,
     y = log10(promoters_ann_core$score),
     main = sprintf("ReMap2020\nSpearman: %f",
                    cor(x = promoters_ann_core$remap_coverage,
                        y = log10(promoters_ann_core$score),
                        method = "spearman")),
     xlab = "promoter coverage",
     ylab = "log10(promoter score)")
graphics::smoothScatter(
     x = promoters_ann_core$chip_atlas_coverage,
     y = log10(promoters_ann_core$score),
     main = sprintf("ChIP-Atlas\nSpearman: %f",
                    cor(x = promoters_ann_core$chip_atlas_coverage,
                        y = log10(promoters_ann_core$score),
                        method = "spearman")),
     xlab = "promoter coverage",
     ylab = "log10(promoter score)")
mtext(
  "Promoter coverage vs promoter score",
  side = 3,
  outer = TRUE,
  line = -2,
  font = 2,
  cex = 1.2
)

# Promoters coverage vs width
graphics::smoothScatter(
     x = promoters_ann_core$remap_coverage,
     y = GenomicRanges::width(promoters_ann_core),
     main = sprintf("ReMap2020\nSpearman: %f",
                    cor(x = promoters_ann_core$remap_coverage,
                        y = GenomicRanges::width(promoters_ann_core),
                        method = "spearman")),
     xlab = "promoter coverage",
     ylab = "promoter width")
graphics::smoothScatter(
     x = promoters_ann_core$chip_atlas_coverage,
     y = GenomicRanges::width(promoters_ann_core),
     main = sprintf("ChIP-Atlas\nSpearman: %f",
                    cor(x = promoters_ann_core$chip_atlas_coverage,
                        y = GenomicRanges::width(promoters_ann_core),
                        method = "spearman")),
     xlab = "promoter coverage",
     ylab = "promoter width")
mtext(
  "Promoter coverage vs promoter width",
  side = 3,
  outer = TRUE,
  line = -2,
  font = 2,
  cex = 1.2
)
```

## Looking at promoter TF interaction matrix column wise (TF wise)

### TF overlapp between ReMap2020 and ChIP-Atlas

```{r include=FALSE}
# ReMap2020
table((remap_meta$tf %in% chip_atlas_meta$tf)) %>%
  `/`(length(remap_meta$tf)) %>% 
  data.frame() %>% 
  dplyr::rename(TF = Var1) %>% 
  dplyr::mutate(TF = c("not included in ChIP-Atlas", "included in ChIP-Atlas")) %>%
  knitr::kable(caption = "ReMap2020 TF")

table(unique(remap_meta$tf) %in% unique(chip_atlas_meta$tf)) %>%
  `/`(length(unique(remap_meta$tf))) %>% 
  data.frame() %>% 
  dplyr::rename(TF = Var1) %>% 
  dplyr::mutate(TF = c("not included in ChIP-Atlas", "included in ChIP-Atlas")) %>%
  knitr::kable(caption = "ReMap2020 unique TF")

remap_meta$tf[! remap_meta$tf %in% chip_atlas_meta$tf] %>%
  table() %>%
  as.numeric() %>%
  plotHistogram(breaks = "Sturges", 
               main = "ReMap2020 TF not included in ChIP-Atlas")

# ChIP-Atlas
table((chip_atlas_meta$tf %in% remap_meta$tf)) %>%
  `/`(length(chip_atlas_meta$tf)) %>% 
  data.frame() %>% 
  dplyr::rename(TF = Var1) %>% 
  dplyr::mutate(TF = c("not included in ReMap2020", "included in ReMap2020")) %>%
  knitr::kable(caption = "ChIP-Atlas TF")

table(unique(chip_atlas_meta$tf) %in% unique(remap_meta$tf)) %>%
  `/`(length(unique(chip_atlas_meta$tf))) %>% 
  data.frame() %>% 
  dplyr::rename(TF = Var1) %>% 
  dplyr::mutate(TF = c("not included in ChIP-Atlas", "included in ChIP-Atlas")) %>%
  knitr::kable(caption = "ChIP-Atlas unique TF")

chip_atlas_meta$tf[! chip_atlas_meta$tf %in% remap_meta$tf] %>%
  table() %>%
  as.numeric() %>%
  plotHistogram(breaks = "Sturges", 
               main = "ChIP-Atlas TF not included in ChIP-Atlas")
```

```{r}
remap_meta[, coverage := Matrix::colSums(remap_promoters_core[, id])]
chip_atlas_meta[, coverage := Matrix::colSums(chip_atlas_promoters_core[, id])]
```

For the basic analysis we have focused only on the ReMap dataset. Choosing only
experiments that overlap more than 1% of core promoters and without any 
condition or treatment label.

Further we focused only on TF with >= 10 experiments and biotypes with >= 29 
experiments, excluding the "breast" and "liver" biotypes.

```{r}
remap_meta_core <- remap_meta[(coverage >= nrow(remap_promoters_core) * 0.01) &
                                (condition == "")]

sel_tf <- remap_meta_core$tf %>% 
  table() %>% 
  `[`(. >= 10) %>% 
  names()
remap_meta_core <- remap_meta_core[tf %in% sel_tf]

sel_biotypes <- remap_meta_core$biotype %>% 
  table() %>% 
  `[`(. >= 29) %>% 
  names() %>%
  `[`(! . %in% c("breast", "liver"))
remap_meta_core <- remap_meta_core[biotype %in% sel_biotypes]

# annotate coverage
remap_meta_core[, cov_type := cut(
  coverage,
  breaks = c(0, quantile(coverage, 0.25), quantile(coverage, 0.75), Inf),
  labels = c("low", "medium", "high")
)]
```

```{r}
plotHistogram(remap_meta_core$coverage, 
              main = "ReMap2020 selected experiment's coverage")
```

### TF / biotype confusion matrix

```{r tf_bg_confusion_matrix, echo=FALSE}
remap_tfs_tab <-
  data.table::dcast(data = remap_meta_core,
                    formula = biotype ~ tf,
                    fun = length)
ord <- c(1, order(remap_tfs_tab[, -1] %>% colSums(), decreasing = TRUE) + 1)
data.table::setcolorder(remap_tfs_tab, ord)
remap_tfs_tab[, .r := -rowSums(remap_tfs_tab[, -1])]
data.table::setorder(remap_tfs_tab, .r)[, .r := NULL]

knitr::kable(remap_tfs_tab[, 1:21], caption = "TF/biotype confusion matrix, top 20 TF")
```

### MYC

```{r}
plotExperimentsHeatmap(
  mat = remap_promoters_core,
  meta = remap_meta_core[tf == "MYC",-c("tf", "condition")])
```

### ELF1

```{r}
plotExperimentsHeatmap(
  mat = remap_promoters_core,
  meta = remap_meta_core[tf == "ELF1",-c("tf", "condition")])
```

### FOXA1

```{r}
plotExperimentsHeatmap(
  mat = remap_promoters_core,
  meta = remap_meta_core[tf == "FOXA1",-c("tf", "condition")])
```

## Experiment collapsing

```{r}
core_id <- colnames(remap_promoters)

remap_promoters_core_tf <- simplifyInteractionMatrix(
  mat = remap_promoters_core,
  gr = remap_meta[id %in% core_id, tf]
)

remap_meta[, tf_biotype := paste0(tf, "_", biotype)]
sel_bio <- remap_meta[id %in% core_id, .N, by = tf_biotype][N >= 3, tf_biotype]
remap_promoters_core_tf_biotype <- simplifyInteractionMatrix(
  mat = remap_promoters_core[, 
             remap_meta[(id %in% core_id) & (tf_biotype %in% sel_bio), id]],
  gr = remap_meta[(id %in% core_id) & (tf_biotype %in% sel_bio), tf_biotype])

promoters_ann_core$remap_tf_coverage <- Matrix::rowSums(remap_promoters_core_tf)
promoters_ann_core$remap_tf_biotype_coverage <-
  Matrix::rowSums(remap_promoters_core_tf_biotype)
with(
  promoters_ann_core,
  graphics::smoothScatter(
    x = remap_promoters_core_tf,
    y = tau,
    main = sprintf(
      "ReMap2020\nSpearman: %f",
      cor(x = remap_promoters_core_tf,
          y = tau,
          method = "spearman")
    ),
    ylab = expression(tau),
    xlab = "promoter coverage"
  )
)
```

## Looking at promoter TF interaction matrix row wise (promoter wise) - lncRNA

Next we have limited our analysis to promoters of lncRNA defined as 
promoters associated with a lncRNA gene (as per GENCODE 38) and supported by 
ENCODE ROADMAP data. Finally, we select one promoter per associated gene that 
have maximum FANTOM5 score.
   
```{r core_lncRNA}
gc()

# only lncRNA genes as per GENCODE annotation
promoters_f5_lncRNA <- 
    promoters_f5[promoters_f5$gene_type_gencode == "lncRNA", ]

# ENCODE ROADMAP confirmation
roadmap_promoters <- rtracklayer::import.bed(
    con = system.file(
        "inst",
        "extdata",
        "Epigenome5DRoadmapDHS_promoter_hg38_liftOver.bed",
        package = "xcore"
    )
)
GenomeInfoDb::seqlevels(roadmap_promoters, pruning.mode = "coarse") <-
    GenomeInfoDb::seqlevels(promoters_f5_lncRNA)
promoters_f5_lncRNA <- intersectGr(promoters_f5_lncRNA, roadmap_promoters)

# Select best promoter per gene
best_promoters <- GenomicRanges::mcols(promoters_f5_lncRNA) %>%
    as.data.frame() %>%
    dplyr::group_by(SYMBOL) %>%
    dplyr::slice(which.max(score)) %>%
    dplyr::pull(name)
promoters_f5_lncRNA <-
    promoters_f5_lncRNA[promoters_f5_lncRNA$name %in% best_promoters, ]

# restrict interaction matrices to core promoters only
remap_promoters_lncRNA <- remap_promoters[promoters_f5_lncRNA$name, ]
chip_atlas_promoters_lncRNA <- chip_atlas_promoters[promoters_f5_lncRNA$name, ]
```

This give us an interaction matrix with `r nrow(remap_promoters_lncRNA)` promoters
(rows) and `r ncol(remap_promoters_lncRNA)`, `r ncol(chip_atlas_promoters_lncRNA)` 
ChIP-seq experiments (columns) for ReMap2020, ChIP-Atlas respectively.

### lncRNA coverage

```{r lncRNA_coverage}
chip_atlas_promoters_lncRNA <- as(chip_atlas_promoters_lncRNA, "dgCMatrix")
lncRNA <- list(
    remap = data.table::data.table(
        nm = remap_promoters_lncRNA %>% rownames(),
        coverage = remap_promoters_lncRNA %>% Matrix::rowSums(),
        tf_coverage = remap_promoters_lncRNA %>% getCoverage(col = 2, sep = "\\."),
        con_coverage = remap_promoters_lncRNA %>% getCoverage(col = 3, sep = "\\.")
    ),
    chip_atlas = data.table::data.table(
        nm = chip_atlas_promoters_lncRNA %>% rownames(),
        coverage = chip_atlas_promoters_lncRNA %>% Matrix::rowSums(),
        tf_coverage = chip_atlas_promoters_lncRNA %>%
            getCoverage(col = 1, sep = "_"),
        con_coverage = chip_atlas_promoters_lncRNA %>%
            getCoverage(col = 3, sep = "_")
    )
)
```

```{r lncRNA_coverage_hist}
# lncRNA coverage histogram
par(mfrow = c(1, 2), cex = 0.7, pty = "s")
coverageHist(x = lncRNA[["remap"]][["coverage"]],
             main = "ReMap2020 lncRNA coverage")
coverageHist(x = lncRNA[["chip_atlas"]][["coverage"]], 
             main = "ChIP-Atlas lncRNA coverage")

# lncRNA TF coverage histogram
par(mfrow = c(1, 2), cex = 0.7, pty = "s")
coverageHist(x = lncRNA[["remap"]][["tf_coverage"]],
             main = "ReMap2020 lncRNA TF coverage")
coverageHist(x = lncRNA[["chip_atlas"]][["tf_coverage"]], 
             main = "ChIP-Atlas lncRNA TF coverage")

# lncRNA biological context coverage histogram
par(mfrow = c(1, 2), cex = 0.7, pty = "s")
coverageHist(x = lncRNA[["remap"]][["con_coverage"]],
             main = "ReMap2020 lncRNA context coverage")
coverageHist(x = lncRNA[["chip_atlas"]][["con_coverage"]], 
             main = "ChIP-Atlas lncRNA context coverage")
```

For some lncRNA we observe no coverage at all:
`r sum(lncRNA[["remap"]][["coverage"]] == 0)`,
`r sum(lncRNA[["chip_atlas"]][["coverage"]] == 0)`
for ReMap2020, ChIP-Atlas respectively.

### ENCODE blacklist

```{r blacklisted_lncRNA}
par(mfrow = c(1, 2), cex = 0.7, pty = "s")
blacklistBoxPlot(coverage = lncRNA[["remap"]][["coverage"]],
                 cov_names = lncRNA[["remap"]][["nm"]],
                 gr1 = blacklist,
                 aname = "blacklisted",
                 bname = "other",
                 main = "ReMap2020\nENCODE blacklisted regions",
                 ylab = "lncRNA coverage")
blacklistBoxPlot(coverage = lncRNA[["chip_atlas"]][["coverage"]],
                 cov_names = lncRNA[["chip_atlas"]][["nm"]],
                 gr1 = blacklist,
                 aname = "blacklisted",
                 bname = "other",
                 main = "ChIP-Atlas\nENCODE blacklisted regions",
                 ylab = "lncRNA coverage")
```

### Ensembl soft masked regions

```{r soft_masked_lncRNA}
par(mfrow = c(1, 2), cex = 0.7, pty = "s")
blacklistBoxPlot(coverage = lncRNA[["remap"]][["coverage"]],
                 cov_names = lncRNA[["remap"]][["nm"]],
                 gr1 = ensembl_sm,
                 aname = "soft masked",
                 bname = "other",
                 main = "ReMap2020\nEnsembl soft masked regions",
                 ylab = "lncRNA coverage")
blacklistBoxPlot(coverage = lncRNA[["chip_atlas"]][["coverage"]],
                 cov_names = lncRNA[["chip_atlas"]][["nm"]],
                 gr1 = ensembl_sm,
                 aname = "soft masked",
                 bname = "other",
                 main = "ChIP-Atlas\nEnsembl soft masked regions",
                 ylab = "lncRNA coverage")
```

### lncRNA promoters "phisical" properties

```{r lncRNA_score_length}
stopifnot(all(promoters_f5_lncRNA$name == lncRNA[["remap"]][["nm"]]))
stopifnot(all(promoters_f5_lncRNA$name == lncRNA[["chip_atlas"]][["nm"]]))

lncRNA_score <- log10(promoters_f5_lncRNA$score)
lncRNA_width <- GenomicRanges::width(promoters_f5_lncRNA)

# lncRNA coverage vs score
par(mfrow = c(1, 2), cex = 0.7, pty = "s")
graphics::smoothScatter(
     x = lncRNA[["remap"]][["coverage"]],
     y = lncRNA_score,
     main = sprintf("ReMap2020\nlncRNA promoter coverage vs F5 score\nSpearman: %f",
                    cor(x = lncRNA[["remap"]][["coverage"]],
                        y = lncRNA_score,
                        method = "spearman")),
     xlab = "lncRNA coverage",
     ylab = "log10(lncRNA score)")
graphics::smoothScatter(
     x = lncRNA[["chip_atlas"]][["coverage"]],
     y = lncRNA_score,
     main = sprintf("ChIP-Atlas\nlncRNA promoter coverage vs F5 score\nSpearman: %f",
                    cor(x = lncRNA[["chip_atlas"]][["coverage"]],
                        y = lncRNA_score,
                        method = "spearman")),
     xlab = "lncRNA promoter coverage",
     ylab = "log10(lncRNA score)")

# lncRNA coverage vs width
par(mfrow = c(1, 2), cex = 0.7, pty = "s")
graphics::smoothScatter(
     x = lncRNA[["remap"]][["coverage"]],
     y = lncRNA_width,
     main = sprintf("ReMap2020\nlncRNA coverage vs width\nSpearman: %f",
                    cor(x = lncRNA[["remap"]][["coverage"]],
                        y = lncRNA_width,
                        method = "spearman")),
     xlab = "lncRNA coverage",
     ylab = "lncRNA width")
graphics::smoothScatter(
     x = lncRNA[["chip_atlas"]][["coverage"]],
     y = lncRNA_width,
     main = sprintf("ChIP-Atlas\nlncRNA coverage vs width\nSpearman: %f",
                    cor(x = lncRNA[["chip_atlas"]][["coverage"]],
                        y = lncRNA_width,
                        method = "spearman")),
     xlab = "lncRNA coverage",
     ylab = "lncRNA width")
```

### FANTOM5 lncRNA specificity

```{r lncRNA_f5_specificity}
promoters_f5_lncRNA_expression <- 
    promoters_f5_expression[`00Annotation` %in% promoters_f5_lncRNA$name, ]
promoters_f5_lncRNA_expression[lncRNA[["remap"]],
                             remap_coverage := coverage,
                             on = c("00Annotation" = "nm")]
promoters_f5_lncRNA_expression[lncRNA[["chip_atlas"]],
                             chip_atlas_coverage := coverage,
                             on = c("00Annotation" = "nm")]
```

```{r lncRNA_f5_tau_hist}
coverageHist(x = promoters_f5_lncRNA_expression$tau,
             main = expression(paste("core lncRNA promoters ", tau, " distribution")))
```

```{r lncRNA_f5_specificity_scatter}
par(mfrow = c(1, 2), cex = 0.7, pty = "s", mgp=c(2,1,0)) # default par(mgp=c(3,1,0))
with(
  promoters_f5_lncRNA_expression,
  graphics::smoothScatter(
    x = remap_coverage,
    y = tau,
    main = sprintf(
      "ReMap2020\nlncRNA specificity by tau\nSpearman: %f",
      cor(x = remap_coverage,
          y = tau,
          method = "spearman")
    ),
    ylab = expression(tau),
    xlab = "lncRNA coverage"
  )
)
with(
  promoters_f5_lncRNA_expression,
  graphics::smoothScatter(
    x = chip_atlas_coverage,
    y = tau,
    main = sprintf(
      "ChIP-Atlas\nlncRNA specificity by tau\nSpearman: %f",
      cor(x = chip_atlas_coverage,
          y = tau,
          method = "spearman")
    ),
    ylab = expression(tau),
    xlab = "lncRNA coverage"
  )
)
```

## Looking at enhancers TF interaction matrix row wise (enhancers wise)

Next we have we look at enhancers defined as enhancers supported by ENCODE 
ROADMAP data.
   
```{r core_enhancers}
gc()

# ENCODE ROADMAP confirmation
roadmap_enhancers <- rtracklayer::import.bed(
    con = system.file(
        "inst",
        "extdata",
        "Epigenome5DRoadmapDHS_enhancer_hg38_liftOver.bed",
        package = "xcore"
    )
)
GenomeInfoDb::seqlevels(roadmap_enhancers, pruning.mode = "coarse") <-
    GenomeInfoDb::seqlevels(xcore::enhancers)
enhancers_core_gr <- intersectGr(xcore::enhancers, roadmap_enhancers)

# additionally we could think of using EP300 or repeat_dfam
# enhancers_core_gr <- enhancers_core_gr[enhancers_core_gr$ep300 == "EP300", ]

# restrict interaction matrices to core promoters only
remap_enhancers_core <- xcore::remap_enhancers[enhancers_core_gr$name, ]
chip_atlas_enhancers_core <- xcore::chip_atlas_enhancers[enhancers_core_gr$name, ]
```

This give us an interaction matrix with `r nrow(remap_enhancers_core)` enhancers
(rows) and `r ncol(remap_enhancers_core)`, `r ncol(chip_atlas_enhancers_core)` 
ChIP-seq experiments (columns) for ReMap2020, ChIP-Atlas respectively.

### Enhancers coverage

```{r enhancers_coverage}
remap_enhancers_core <- as(remap_enhancers_core, "dgCMatrix")
chip_atlas_enhancers_core <- as(chip_atlas_enhancers_core, "dgCMatrix")

enhancers_core <- list(
    remap = data.table::data.table(
        nm = remap_enhancers_core %>% rownames(),
        coverage = remap_enhancers_core %>% Matrix::rowSums(),
        tf_coverage = remap_enhancers_core %>% getCoverage(col = 2, sep = "\\."),
        con_coverage = remap_enhancers_core %>% getCoverage(col = 3, sep = "\\.")
    ),
    chip_atlas = data.table::data.table(
        nm = chip_atlas_enhancers_core %>% rownames(),
        coverage = chip_atlas_enhancers_core %>% Matrix::rowSums(),
        tf_coverage = chip_atlas_enhancers_core %>%
            getCoverage(col = 1, sep = "_"),
        con_coverage = chip_atlas_enhancers_core %>%
            getCoverage(col = 3, sep = "_")
    )
)
```

```{r enhancers_coverage_hist}
# enhancers coverage histogram
par(mfrow = c(1, 2), cex = 0.7, pty = "s")
coverageHist(x = enhancers_core[["remap"]][["coverage"]],
             main = "ReMap2020 enhancers coverage")
coverageHist(x = enhancers_core[["chip_atlas"]][["coverage"]], 
             main = "ChIP-Atlas enhancers coverage")

# enhancers TF coverage histogram
par(mfrow = c(1, 2), cex = 0.7, pty = "s")
coverageHist(x = enhancers_core[["remap"]][["tf_coverage"]],
             main = "ReMap2020 enhancers TF coverage")
coverageHist(x = enhancers_core[["chip_atlas"]][["tf_coverage"]], 
             main = "ChIP-Atlas enhancers TF coverage")

# enhancers biological context coverage histogram
par(mfrow = c(1, 2), cex = 0.7, pty = "s")
coverageHist(x = enhancers_core[["remap"]][["con_coverage"]],
             main = "ReMap2020 enhancers context coverage")
coverageHist(x = enhancers_core[["chip_atlas"]][["con_coverage"]], 
             main = "ChIP-Atlas enhancers context coverage")
```

For some enhancers we observe no coverage at all:
`r sum(enhancers_core[["remap"]][["coverage"]] == 0)`,
`r sum(enhancers_core[["chip_atlas"]][["coverage"]] == 0)`
for ReMap2020, ChIP-Atlas respectively.

### ENCODE blacklist

```{r blacklisted_enhancers}
par(mfrow = c(1, 2), cex = 0.7, pty = "s")
blacklistBoxPlot(coverage = enhancers_core[["remap"]][["coverage"]],
                 cov_names = enhancers_core[["remap"]][["nm"]],
                 gr1 = blacklist,
                 aname = "blacklisted",
                 bname = "other",
                 main = "ReMap2020\nENCODE blacklisted regions",
                 ylab = "enhancers coverage")
blacklistBoxPlot(coverage = enhancers_core[["chip_atlas"]][["coverage"]],
                 cov_names = enhancers_core[["chip_atlas"]][["nm"]],
                 gr1 = blacklist,
                 aname = "blacklisted",
                 bname = "other",
                 main = "ChIP-Atlas\nENCODE blacklisted regions",
                 ylab = "enhancers coverage")
```

### Ensembl soft masked regions

```{r soft_masked_enhancers}
par(mfrow = c(1, 2), cex = 0.7, pty = "s")
blacklistBoxPlot(coverage = enhancers_core[["remap"]][["coverage"]],
                 cov_names = enhancers_core[["remap"]][["nm"]],
                 gr1 = ensembl_sm,
                 aname = "soft masked",
                 bname = "other",
                 main = "ReMap2020\nEnsembl soft masked regions",
                 ylab = "enhancers coverage")
blacklistBoxPlot(coverage = enhancers_core[["chip_atlas"]][["coverage"]],
                 cov_names = enhancers_core[["chip_atlas"]][["nm"]],
                 gr1 = ensembl_sm,
                 aname = "soft masked",
                 bname = "other",
                 main = "ChIP-Atlas\nEnsembl soft masked regions",
                 ylab = "enhancers coverage")
```

### enhancers "phisical" properties

```{r enhancers_score_length}
stopifnot(all(enhancers_core_gr$name == enhancers_core[["remap"]][["nm"]]))
stopifnot(all(enhancers_core_gr$name == enhancers_core[["chip_atlas"]][["nm"]]))

# enhancers_score <- log10(enhancers_core_gr$score)
enhancers_width <- GenomicRanges::width(enhancers_core_gr)

# enhancers coverage vs width
par(mfrow = c(1, 2), cex = 0.7, pty = "s")
graphics::smoothScatter(
     x = enhancers_core[["remap"]][["coverage"]],
     y = enhancers_width,
     main = sprintf("ReMap2020\nenhancer coverage vs width\nSpearman: %f",
                    cor(x = enhancers_core[["remap"]][["coverage"]],
                        y = enhancers_width,
                        method = "spearman")),
     xlab = "enhancer coverage",
     ylab = "enhancer width")
graphics::smoothScatter(
     x = enhancers_core[["chip_atlas"]][["coverage"]],
     y = enhancers_width,
     main = sprintf("ChIP-Atlas\nenhancer coverage vs width\nSpearman: %f",
                    cor(x = enhancers_core[["chip_atlas"]][["coverage"]],
                        y = enhancers_width,
                        method = "spearman")),
     xlab = "enhancer coverage",
     ylab = "enhancer width")
```

### FANTOM5 enhancer specificity

```{r enhancers_f5_specificity}
enhancers_f5_core_expression <- 
    enhancers_f5_expression[`V1` %in% enhancers_core_gr$name, ]
enhancers_f5_core_expression[enhancers_core[["remap"]],
                             remap_coverage := coverage,
                             on = c("V1" = "nm")]
enhancers_f5_core_expression[enhancers_core[["chip_atlas"]],
                             chip_atlas_coverage := coverage,
                             on = c("V1" = "nm")]
```

```{r enhancers_f5_tau_hist}
coverageHist(x = enhancers_f5_core_expression$tau,
             main = expression(paste("core enhancers ", tau, " distribution")))
```

```{r enhancers_f5_specificity_scatter}
par(mfrow = c(1, 2), cex = 0.7, pty = "s", mgp=c(2,1,0)) # default par(mgp=c(3,1,0))
with(
  enhancers_f5_core_expression,
  graphics::smoothScatter(
    x = remap_coverage,
    y = tau,
    main = sprintf(
      "ReMap2020\nenhancer specificity by tau\nSpearman: %f",
      cor(x = remap_coverage,
          y = tau,
          method = "spearman")
    ),
    ylab = expression(tau),
    xlab = "enhancer coverage"
  )
)
with(
  enhancers_f5_core_expression,
  graphics::smoothScatter(
    x = chip_atlas_coverage,
    y = tau,
    main = sprintf(
      "ChIP-Atlas\nenhancer specificity by tau\nSpearman: %f",
      cor(x = chip_atlas_coverage,
          y = tau,
          method = "spearman")
    ),
    ylab = expression(tau),
    xlab = "enhancer coverage"
  )
)
```

## Looking at promoter TF interaction matrix column wise (TF wise)

```{r tf_coverage}
# Functions
# Function calculates between columns correlation over sparse matrix
# Taken from https://stackoverflow.com/questions/5888287/running-cor-or-any-variant-over-a-sparse-matrix-in-r
sparse.cor <- function(x){
    n <- nrow(x)

    cMeans <- Matrix::colMeans(x)
    cSums <- Matrix::colSums(x)

    # Calculate the population covariance matrix.
    # There's no need to divide by (n-1) as the std. dev is also calculated the same way.
    # The code is optimized to minize use of memory and expensive operations
    covmat <- Matrix::tcrossprod(cMeans, (-2 * cSums + n * cMeans))
    crossp <- Matrix::as.matrix(Matrix::crossprod(x))
    covmat <- covmat + crossp

    sdvec <- sqrt(Matrix::diag(covmat)) # standard deviations of columns
    covmat / Matrix::crossprod(Matrix::t(sdvec)) # correlation matrix
}
```

There are `r sum(remap_tfs$coverage == 0)`, `r sum(chip_atlas_tfs$coverage == 0)`
zero coverage ChIP-seqs in ReMap2020, ChIP-Atlas respectively. I've removed 
those in following analyses.

Moreover in ChIP-Atlas there is `r sum(chip_atlas_tfs$biotype == "Unclassified)`
with 'Unclassified' biotype, those are removed as well.

```{r chip_cleanup}
remap_tfs <- remap_tfs[remap_tfs$coverage != 0]
chip_atlas_tfs <- chip_atlas_tfs[chip_atlas_tfs$coverage != 0]
chip_atlas_tfs <- chip_atlas_tfs[chip_atlas_tfs$biotype != "Unclassified"]
```

```{r chip_seq_coverage}
par(mfrow = c(1, 2), cex = 0.7, pty = "s")
coverageHist(x = remap_tfs$coverage,
             main = "ReMap2020 ChIP-seq coverage")
coverageHist(x = chip_atlas_tfs$coverage, 
             main = "ChIP-Atlas ChIP-seq coverage")

coverageHist(x = remap_tfs$tf %>% table() %>% as.numeric(),
             main = "ReMap2020 TFs")
coverageHist(x = chip_atlas_tfs$tf %>% table() %>% as.numeric(), 
             main = "ChIP-Atlas TFs")

coverageHist(x = remap_tfs$biotype %>% table() %>% as.numeric(),
             main = "ReMap2020 biotypes")
coverageHist(x = chip_atlas_tfs$biotype %>% table() %>% as.numeric(), 
             main = "ChIP-Atlas biotypes")
```

### TF overlapp between ReMap2020 and ChIP-Atlas

```{r}
# ReMap2020
table((remap_tfs$tf %in% chip_atlas_tfs$tf)) %>%
  `/`(length(remap_tfs$tf)) %>% 
  data.frame() %>% 
  dplyr::rename(TF = Var1) %>% 
  dplyr::mutate(TF = c("not included in ChIP-Atlas", "included in ChIP-Atlas")) %>%
  knitr::kable(caption = "ReMap2020 TF")

table(unique(remap_tfs$tf) %in% unique(chip_atlas_tfs$tf)) %>%
  `/`(length(unique(remap_tfs$tf))) %>% 
  data.frame() %>% 
  dplyr::rename(TF = Var1) %>% 
  dplyr::mutate(TF = c("not included in ChIP-Atlas", "included in ChIP-Atlas")) %>%
  knitr::kable(caption = "ReMap2020 unique TF")

remap_tfs$tf[! remap_tfs$tf %in% chip_atlas_tfs$tf] %>%
  table() %>%
  hist(main = "ReMap2020 TF not included in ChIP-Atlas")

# ChIP-Atlas
table((chip_atlas_tfs$tf %in% remap_tfs$tf)) %>%
  `/`(length(chip_atlas_tfs$tf)) %>% 
  data.frame() %>% 
  dplyr::rename(TF = Var1) %>% 
  dplyr::mutate(TF = c("not included in ReMap2020", "included in ReMap2020")) %>%
  knitr::kable(caption = "ChIP-Atlas TF")

table(unique(chip_atlas_tfs$tf) %in% unique(remap_tfs$tf)) %>%
  `/`(length(unique(chip_atlas_tfs$tf))) %>% 
  data.frame() %>% 
  dplyr::rename(TF = Var1) %>% 
  dplyr::mutate(TF = c("not included in ChIP-Atlas", "included in ChIP-Atlas")) %>%
  knitr::kable(caption = "ChIP-Atlas unique TF")

chip_atlas_tfs$tf[! chip_atlas_tfs$tf %in% remap_tfs$tf] %>%
  table() %>%
  hist(main = "ChIP-Atlas TF not included in ChIP-Atlas")
```

### TF / background confusion matrix

```{r tf_bg_confusion_matrix}
remap_tfs_tab <-
  data.table::dcast(data = remap_tfs,
                    formula = biotype ~ tf,
                    fun = length)
ord <- c(1, order(remap_tfs_tab[, -1] %>% colSums(), decreasing = TRUE) + 1)
data.table::setcolorder(remap_tfs_tab, ord)
remap_tfs_tab[, .r := -rowSums(remap_tfs_tab[, -1])]
data.table::setorder(remap_tfs_tab, .r)[, .r := NULL]

chip_atlas_tfs_tab <-
  data.table::dcast(data = chip_atlas_tfs,
                    formula = biotype ~ tf,
                    fun = length)
ord <- c(1, order(chip_atlas_tfs_tab[, -1] %>% colSums(), decreasing = TRUE) + 1)
data.table::setcolorder(chip_atlas_tfs_tab, ord)
chip_atlas_tfs_tab[, .r := -rowSums(chip_atlas_tfs_tab[, -1])]
data.table::setorder(chip_atlas_tfs_tab, .r)[, .r := NULL]
```
### Case study: select TP53, MYC other similar specific TFs and show heatmap
### with clustering, does they cluster by study or by experiment? -- towards
### collapsing method

# TP53

```{r tp53}
remap_tfs[tf == "TP53", 
          .(
            n_id = length(id),
            n_study = length(unique(study)),
            n_biotype = length(unique(biotype))
           )]

remap_tfs[tf == "TP53", 
          hist(table(study), main = "TP53 study")
          ]
remap_tfs[tf == "TP53", 
          hist(table(biotype), main = "TP53 biotype")
          ]

m <- grepl("\\.TP53\\.", colnames(remap_promoters_core))
tp53_remap <- remap_promoters_core[, m]

coverageHist(Matrix::colSums(tp53_remap), main = "TP53 coverage")

# remove 0 coverage experiments
tp53_remap <- tp53_remap[, Matrix::colSums(tp53_remap) > 0]

tp53_remap_cor <- sparse.cor(tp53_remap)
tp53_d <- as.dist(abs((1 - tp53_remap_cor) / 2))
tp53_d <- dist(Matrix::t(tp53_remap))
tp53_tree <- hclust(tp53_d, method = "complete")
plot(tp53_tree, hang = -1, cex = 0.5)

study_labs <- gsub("\\.TP53.*", "", tp53_tree$labels)[tp53_tree$order]
study_grp <- study_labs %>% as.factor() %>% unclass()
biotype_cond_labs <- gsub(".*.TP53\\.", "", tp53_tree$labels)[tp53_tree$order]
biotype_cond_grp <- biotype_cond_labs %>% as.factor() %>% unclass()
biotype_labs <- gsub("_.*", "", biotype_cond_labs)
biotype_grp <- biotype_labs %>% as.factor() %>% unclass()

getPch <- function(labs) {
  grp <- labs %>% as.factor() %>% unclass()
  pch <- rep(0:18, ceiling(max(grp) / 19))
  pch[grp]
}
getCol <- function(labs, cols=cols) {
  grp <- labs %>% as.factor() %>% unclass()
  cols <- rep(c("red", "yellow", "blue", "orange", "green", "violet"), 19)
  cols[grp]
}

# study
stats::as.dendrogram(tp53_tree) %>% 
  dendextend::set("labels", biotype_cond_labs) %>%
  dendextend::set("labels_cex", 0.5) %>%
  dendextend::set("leaves_pch", getPch(study_labs)) %>%
  dendextend::set("leaves_col", getCol(study_labs)) %>% 
  plot(dLeaf = 0.02)

# biotype_cond
stats::as.dendrogram(tp53_tree) %>% 
  dendextend::set("labels", study_labs) %>%
  dendextend::set("labels_cex", 0.5) %>%
  dendextend::set("leaves_pch", getPch(biotype_cond_labs)) %>%
  dendextend::set("leaves_col", getCol(biotype_cond_labs)) %>% 
  plot(dLeaf = 0.02)

# biotype
stats::as.dendrogram(tp53_tree) %>% 
  dendextend::set("labels", biotype_labs) %>%
  dendextend::set("labels_cex", 0.5) %>%
  dendextend::set("leaves_pch", getPch(biotype_labs)) %>%
  dendextend::set("leaves_col", getCol(biotype_labs)) %>% 
  plot(dLeaf = 0.02)

# ape
tp53_phylo <- ape::as.phylo(tp53_tree)

# study
plot(tp53_phylo, 
     type = "fan",
     label.offset = 0.02, 
     cex = 0.7)
tiplabels(pch = getPch(tp53_phylo$tip.label %>% gsub("\\.TP53.*", "", .)), 
          col = getCol(tp53_phylo$tip.label %>% gsub("\\.TP53.*", "", .)))

# biotype
plot(tp53_phylo, 
     type = "fan",
     label.offset = 0.02, 
     cex = 0.7)
tiplabels(pch = getPch(tp53_phylo$tip.label %>% 
                         gsub(".*TP53\\.", "", .) %>% 
                         gsub("_.*", "", .)), 
          col = getCol(tp53_phylo$tip.label %>% 
                         gsub(".*TP53\\.", "", .) %>% 
                         gsub("_.*", "", .)))

# study / biotype
getCol <- function(labs, cols=cols) {
  grp <- labs %>% as.factor() %>% unclass()
  cols <- rainbow(max(grp))
  cols[grp]
}
plot(tp53_phylo, 
     type = "fan",
     label.offset = 0.02, 
     cex = 0.7)
tiplabels(pch = getPch(tp53_phylo$tip.label %>% 
                         gsub(".*TP53\\.", "", .) %>% 
                         gsub("_.*", "", .)), 
          col = getCol(tp53_phylo$tip.label %>% gsub("\\.TP53.*", "", .)))

study_ann <- data.frame(
  study = tp53_remap_cor %>%
    rownames() %>%
    gsub("\\.TP53.*", " ", .),
  row.names = tp53_remap_cor %>% rownames(),
  stringsAsFactors = FALSE
)
biotype_ann <- data.frame(
  biotype = tp53_remap_cor %>%
    rownames() %>%
    gsub(".*TP53\\.", "", .) %>%
    gsub("_.*", "", .),
  row.names = tp53_remap_cor %>% rownames(),
  stringsAsFactors = FALSE
)
annotation_colors <- list(
  study = stats::setNames(
    pals::glasbey(study_ann$study %>% as.factor() %>% nlevels()),
    study_ann$study %>% as.factor() %>% levels()
  ),
  biotype = stats::setNames(
    pals::glasbey(biotype_ann$biotype %>% as.factor() %>% nlevels()),
    biotype_ann$biotype %>% as.factor() %>% levels()
  )
)
pheatmap::pheatmap(mat = tp53_remap_cor, 
                   scale = "none", 
                   cluster_rows = TRUE,
                   cluster_cols = TRUE, 
                   clustering_distance_rows = as.dist(abs(1 - tp53_remap_cor) / 2),
                   clustering_distance_cols = as.dist(abs(1 - tp53_remap_cor) / 2),
                   fontsize = 4,
                   labels_col = tp53_remap_cor %>%
                     rownames() %>% 
                     gsub("\\.TP53\\.", " ", .),
                   labels_row = tp53_remap_cor %>%
                     rownames() %>% 
                     gsub("\\.TP53\\.", " ", .),
                   annotation_row = study_ann,
                   annotation_col = biotype_ann, 
                   annotation_colors = annotation_colors
)
```

```{r tf_coverage_exploration_remap}

# study
object <- lm(coverage ~ term, 
             data = data.table::dcast(data = remap_tfs, 
                                      formula = id + coverage + term ~ study, 
                                      fun = length))
res_study <- midasHLA::analyzeAssociations(object = object,
                                           variables = unique(remap_tfs$study), 
                                           correction = "BH")

# tf
object <- lm(coverage ~ term, 
             data = data.table::dcast(data = remap_tfs, 
                                      formula = id + coverage + term ~ tf, 
                                      fun = length))
res_tf <- midasHLA::analyzeAssociations(object = object,
                                           variables = unique(remap_tfs$tf), 
                                           correction = "BH")

# biotype
object <- lm(coverage ~ term, 
             data = data.table::dcast(data = remap_tfs, 
                                      formula = id + coverage + term ~ biotype, 
                                      fun = length))
res_biotype <- midasHLA::analyzeAssociations(object = object,
                                           variables = unique(remap_tfs$biotype), 
                                           correction = "BH")
```

For few of the studies we can observe tendencies for ChIP-seq experiment to 
have high or low coverage.

In ReMap2020 we can find 3 large such studies:

#### GSE78099 
consisting out of 208 experiments that has a tendency towards low coverage.

```{r}
m <- remap_tfs$study == "GSE78099"
graphics::boxplot(remap_tfs$coverage[m], remap_tfs$coverage[!m], 
                  names = c("GSE78099", "other"))
```

This study has many different TFs, mainly different ZNFs in HEK293T background.
So of those are also present in other studies so we can compare to those.

```{r}
GSE78099_tfs <- remap_tfs$tf[remap_tfs$study == "GSE78099"]
remap_tfs$tf[remap_tfs$tf %in% GSE78099_tfs] %>% 
  table() %>% 
  `[`(. > 1) %>%
  names() -> tf_sel

m1 <- remap_tfs$study == "GSE78099"
m2 <- remap_tfs$tf %in% tf_sel
graphics::boxplot(remap_tfs$coverage[m1 & m2], remap_tfs$coverage[(! m1) & m2], 
                  names = c("GSE78099", "other"),
                  main = "Selected TFs")
```

To make it even more comparable we will restrict ourselves to HEK293 and HEK293T
biotypes.

```{r}
m1 <- remap_tfs$study == "GSE78099"  
m2 <- remap_tfs$tf %in% tf_sel
m3 <- remap_tfs$biotype %in% c("HEK293", "HEK293T")
graphics::boxplot(remap_tfs$coverage[m1 & m2 & m3], remap_tfs$coverage[(! m1) & m2 &m3], 
                  names = c("GSE78099", "other"),
                  main = "Selected TFs",
                  outline = FALSE)
```

Still it looks like the coverage in GSE78099 is lower then in other studies.

####GSE87418
is a high coverage study with a smaller number of samples = 36.

Here they explore BRD4, CEBPB, MED1 in HCC1806, SUM159PT or SUM229PE under
different conditions.

```{r}
m <- remap_tfs$study == "GSE87418"
graphics::boxplot(remap_tfs$coverage[m], remap_tfs$coverage[!m], 
                  names = c("GSE87418", "other"))
```

Again let's try comparing with other experiments

```{r}
GSE87418_tfs <- remap_tfs$tf[remap_tfs$study == "GSE87418"]
remap_tfs$tf[remap_tfs$tf %in% GSE87418_tfs] %>% 
  table() %>% 
  `[`(. > 1) %>%
  names() -> tf_sel

m1 <- remap_tfs$study == "GSE87418"
m2 <- remap_tfs$tf %in% tf_sel
graphics::boxplot(remap_tfs$coverage[m1 & m2], remap_tfs$coverage[(! m1) & m2], 
                  names = c("GSE87418", "other"),
                  main = "Selected TFs")
```

####GSE109411
is a high coverage study with 12 samples.

Here again we have BRD4 TF in different conditions.So let's just compare with
GSE87418 BRD samples.

```{r}
m1 <- remap_tfs$study == "GSE87418"
m2 <- remap_tfs$study == "GSE109411"
m3 <- remap_tfs$tf == "BRD4"
graphics::boxplot(remap_tfs$coverage[m1 & m3], remap_tfs$coverage[m2 & m3], 
                  names = c("GSE87418", "GSE109411"),
                  main = "BRD")
```

The above result would actually advocate that there is no study bias but rather
TF bias ie. BRD4 has many binding sites in promoters. 

This result also underlines the need to control for TF and likely biotype when
modeling study bias. But how to do this?

####GSE100292 
is a low coverage study with a smaller number of samples = 17

here they explored TP53 across different biotypes and always with some kind of
treatment. I've looked up thier paper and they also report small number of 
detected peaks.

```{r}
m <- remap_tfs$study == "GSE100292"
graphics::boxplot(remap_tfs$coverage[m], remap_tfs$coverage[!m], 
                  names = c("GSE100292", "other"))
```

Again let's try comparing with tp53 from other experiments

```{r}
GSE100292_tfs <- remap_tfs$tf[remap_tfs$study == "GSE100292"]
remap_tfs$tf[remap_tfs$tf %in% GSE100292_tfs] %>% 
  table() %>% 
  `[`(. > 1) %>%
  names() -> tf_sel

m1 <- remap_tfs$study == "GSE100292"
m2 <- remap_tfs$tf %in% tf_sel
graphics::boxplot(remap_tfs$coverage[m1 & m2], remap_tfs$coverage[(! m1) & m2], 
                  names = c("GSE100292", "other"),
                  main = "Selected TFs",
                  outline = FALSE)
```

####GSE101074
is a high coverage study with small number of experiments = 3

Here they explore TFAP2C in UCLA1-hESCs under different conditions.

```{r}
m <- remap_tfs$study == "GSE101074"
graphics::boxplot(remap_tfs$coverage[m], remap_tfs$coverage[!m], 
                  names = c("GSE101074", "other"))
```

Again let's try comparing with other experiments

```{r}
GSE101074_tfs <- remap_tfs$tf[remap_tfs$study == "GSE101074"]
remap_tfs$tf[remap_tfs$tf %in% GSE101074_tfs] %>% 
  table() %>% 
  `[`(. > 1) %>%
  names() -> tf_sel

m1 <- remap_tfs$study == "GSE101074"
m2 <- remap_tfs$tf %in% tf_sel
graphics::boxplot(remap_tfs$coverage[m1 & m2], remap_tfs$coverage[(! m1) & m2], 
                  names = c("GSE101074", "other"),
                  main = "Selected TFs")
```

```{r tf_coverage_exploration_chipatlas}
# study
object <- lm(coverage ~ term, 
             data = data.table::dcast(data = chip_atlas_tfs, 
                                      formula = id + coverage + term ~ id, 
                                      fun = length))
chip_atlas_res_study <- midasHLA::analyzeAssociations(object = object,
                                           variables = unique(chip_atlas_tfs$id), 
                                           correction = "BH")

# tf
object <- lm(coverage ~ term, 
             data = data.table::dcast(data = chip_atlas_tfs, 
                                      formula = id + coverage + term ~ tf, 
                                      fun = length))
chip_atlas_res_tf <- midasHLA::analyzeAssociations(object = object,
                                           variables = unique(chip_atlas_tfs$tf), 
                                           correction = "BH")

# biotype
object <- lm(coverage ~ term, 
             data = data.table::dcast(data = chip_atlas_tfs, 
                                      formula = id + coverage + term ~ biotype, 
                                      fun = length))
chip_atlas_res_biotype <- midasHLA::analyzeAssociations(object = object,
                                           variables = unique(chip_atlas_tfs$biotype), 
                                           correction = "BH")
```


# what are the chip seqs with least/most peaks? How do they compare to average
# ChIPs?

```{r looking_into_datasets}
getFeature <- function(nm, col, sep = "\\.") {
    n <- stringr::str_count(nm, sep)[1]
    nm %>% 
        gsub(pattern = paste(c("(.*)", rep(c(sep, "(.*)"), n)), collapse = ""),
             replacement = paste0("\\", col))
}

remap_tfs <- remap_promoters_core %>%
  colnames() %>%
  stringr::str_split(pattern = "\\.", n = 3) %>%
  do.call(what = rbind) %>%
  data.table::as.data.table()
colnames(remap_tfs) <- c("id", "tf", "background")

chip_atlas_tfs <- chip_atlas_promoters_core %>%
  colnames() %>%
  stringr::str_split(pattern = "_", n = 4) %>%
  do.call(what = rbind) %>%
  data.table::as.data.table()
colnames(chip_atlas_tfs) <- c("tf", "context", "background", "id")

table((remap_tfs$background %>% unique()) %in% (chip_atlas_tfs$background %>% unique()))
remap_tfs$background %>% unique() %>% length()

table((chip_atlas_tfs$background %>% unique()) %in% (remap_tfs$background %>% unique()))
chip_atlas_tfs$background %>% unique() %>% length()
```

# Can we summarize redundant signatures to e.g. one per TF, TF in cell type
## replicated expriments in ChIP-Atlas

# Concordance between signatures of the same TF/same cell line. (difference 
# between studies, similarities within studies)

