---
title: "xcore vignette"
author: "Migdal"
date: "12/2021"
output: html_document
vignette: >
  %\VignetteIndexEntry{xcore vignette}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---


```{r include=FALSE}
knitr::opts_chunk$set(warning = FALSE, message = FALSE)

devtools::load_all()
```

This document introduces the xcore package and showcase its basic usage.

# Gene expression modeling in context of rinderpest infection

Here we will use subset of 293SLAM rinderpest infection dataset from
\link[FANTOM5]{https://fantom.gsc.riken.jp/5/sstar/Rinderpest_infection_series}.
This subset contains expression counts for 0, 12 and 24 hours post infection
samples and only for a subset of FANTOM5 promoters. We can find this example
data shipped with \code{xcore} package.

```{r}
head(rinderpest_mini)
```

First we need to construct a design matrix describing our experiment design.

```{r}
design <- matrix(
  data = c(1, 0, 0,
           1, 0, 0,
           1, 0, 0,
           0, 1, 0,
           0, 1, 0,
           0, 1, 0,
           0, 0, 1,
           0, 0, 1,
           0, 0, 1),
  ncol = 3,
  nrow = 9,
  byrow = TRUE,
  dimnames = list(
    c(
      "00hr_rep1",
      "00hr_rep2",
      "00hr_rep3",
      "12hr_rep1",
      "12hr_rep2",
      "12hr_rep3",
      "24hr_rep1",
      "24hr_rep2",
      "24hr_rep3"
    ),
    c("00hr", "12hr", "24hr")
  )
)

print(design)
```

Next, we need to preprocess the counts using \code{prepareCountsForRegression}
function. Here, CAGE expression tags for each sample are filtered for lowly
expressed promoters, normalized for the library size and transformed into counts
per million (CPM). Finally, CPM are log2 transformed with addition of pseudo
count 1. Moreover, we designate the base level samples, 0 hours after treatment
in our example, from which basal expression level is calculated. This basal
level will be used as a reference when modeling expression changes.

```{r}
mae <- prepareCountsForRegression(counts = rinderpest_mini,
                                  design = design,
                                  base_lvl = "00hr")
```

\code{xcore} models the expression as a function of molecular signatures. Such
signatures can be constructed eg. from the known transcription factor binding
sites (see [Constructing molecular signatures] section). Here, we will take
advantage of pre-computed molecular signatures found in \code{xcoredata} package.
Particularly, molecular signatures constructed from ReMap2020 against FANTOM5
annotation.

```{r}
data(remap_promoters, package = "xcoredata")
```

Molecular signature is a simple binary matrix indicating if transcription factor
binding site was found or not in particular promoter vicinity.

```{r}
print(remap_promoters[1:3, 1:3])
```

To add signatures to our \code{MultiAssayExperiment} object we can use
\code{addSignatures} function. As you add your signatures remember to give them
unique names.

```{r}
mae <- addSignatures(mae, remap = remap_promoters)
```

When we examine newly added signatures, we can see that some of them does not
overlap any of the promoters. On the other side, depending on the signatures
quality, we could expect to see signatures that overlap all of the promoters.

```{r}
summary(colSums(mae[["remap"]]))
```

While, we do not provide detailed guidelines to signatures filtering, filtering
out at least signatures overlapping no or all promoters is mandatory.
Here, we filter signatures that overlap less than 5% or more than 95% of
promoters using \code{filterSignatures} function.

```{r}
mae <- filterSignatures(mae, min = 0.05, max = 0.95)
```

Finally, we can run our expression modeling using \code{modelGeneExpression}
function. \code{modelGeneExpression} can take advantage of parallelization to
speedup its calculations, to use it we need to register parallel backend.

Our modeling will inherently contain some level of randomness due to
using cross validation, here we set the seed so that our results can be
replicated.

This step is time consuming as we need to train a separate model for each of our
samples. To speed up the calculations we will lower number of folds used in
cross-validation procedure from 10 to 5 using \code{nfolds} argument, which is
internally passed to \code{cv.glmnet} function. Using 6 cores on
Intel(R) Xeon(R) CPU E5-2680 v3 calculations took around 4 minutes to compute
and used up to 25 GB RAM. If needed memory requirement could be brought down by
using lower number of cores.

```{r}
# register parallel backend
doMC::registerDoMC(cores = 6L)

# set seed
set.seed(314159265)

res <- modelGeneExpression(
  mae = mae,
  xnames = "remap",
  nfolds = 5)
```


Output returned by \code{modelGeneExpression} is a list with following elements:

+ \code{regression_models} - a list holding \code{cv.glmnet} objects for each
  sample.
+ \code{pvalues} - list of \code{data.frame}s for each sample, each holding
  signatures estimates, their estimated standard errors and p-values.
+ \code{replicate_avg} - a matrix holding replicate average Z-scores with
  columns corresponding to groups in the design. This is likely the main output
  for most of the users.

Let's have a look at \code{replicate_avg} matrix.

```{r}
out <- res$replicate_avg$remap
ord <- order(abs(out[, "24hr"]), decreasing = TRUE)
pheatmap::pheatmap(
  mat = head(out[ord, ]),
  cluster_cols = FALSE,
  main = "SLAM293 signatures Z-score"
)
```

Among others, our analysis indicates RELA and IRF9 as the top transcription
factor associated with observed changes in expression. These transcription
factors are known to play a role during measles infection, which is closely
related to rinderpest
(\link[KEGG Measles pathway]{https://www.genome.jp/kegg-bin/show_pathway?hsa05162}).

Moreover, Z-score sign indicate that promoters targeted by RELA tend to be
downregulated, while those targeted by IRF9 tend to be upregulated.


# Constructing molecular signatures

Constructing molecular signatures is as simple as intersecting promoters
annotation with TFBS. For example here we downloaded Motevo predictions of
binding sites from SwissRegulon website. The file is GFF format and we can
read it using \code{rtracklayer::import}. After we read it into R we will get
a \code{GRanges} object. For our purposes it is required that TF/motif name is
held under \code{name} column.

Next we construct our molecular signatures matrix using FANTOM5 annotation found
in \code{xcoredata} package. This is done using \code{getInteractionMatrix}
function, where we can specify the \code{ext} parameter that will control how
much promoters are extended in both directions before intersecting with binding
data.

```{r}
# obtain binding sites
motevo_hg38_f5 <- rtracklayer::import(
  con = "https://swissregulon.unibas.ch/data/hg38_f5/hg38_sites_v1.gff.gz"
)

# TF / motif name must be stored as 'name'
motevo_hg38_f5$name <- motevo_hg38_f5$Motif

# obtain promoters annotation, *promoter name must be stored as 'name'
data("promoters_f5", package = "xcoredata")

# construct molecular signatures matrix
molecular_signature <- getInteractionMatrix(
  a = promoters_f5,
  b = motevo_hg38_f5,
  ext = 500
)
```
