<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="xcore">
<title>xcore vignette • xcore</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="xcore vignette">
<meta property="og:description" content="xcore">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">xcore</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9000</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-2">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/xcore_vignette.html">xcore vignette</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-article">
<script src="xcore_vignette_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <main id="main"><div class="page-header">
      <img src="" class="logo" alt=""><h1>xcore vignette</h1>
                        <h4 data-toc-skip class="author">Migdal</h4>
            
            <h4 data-toc-skip class="date">12/2021</h4>
      
      
      <div class="d-none name"><code>xcore_vignette.Rmd</code></div>
    </div>

    
    
<p>This document introduces the xcore package and showcase its basic usage.</p>
<div class="section level1">
<h1 id="gene-expression-modeling-in-context-of-rinderpest-infection">Gene expression modeling in context of rinderpest infection<a class="anchor" aria-label="anchor" href="#gene-expression-modeling-in-context-of-rinderpest-infection"></a>
</h1>
<p>Here we will use subset of 293SLAM rinderpest infection dataset from <a href="https://fantom.gsc.riken.jp/5/sstar/Rinderpest_infection_series" class="external-link">FANTOM5</a>. This subset contains expression counts for 0, 12 and 24 hours post infection samples and only for a subset of FANTOM5 promoters. We can find this example data shipped with <code>xcore</code> package.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">rinderpest_mini</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">##                                              00hr_rep1 00hr_rep2 00hr_rep3</span>
<span class="co">## hg19::chr1:10003372..10003465,-;hg_10258.1          52        46        57</span>
<span class="co">## hg19::chr1:10003486..10003551,+;hg_541.1            39        42        27</span>
<span class="co">## hg19::chr1:100111580..100111773,+;hg_4181.1          1         0         2</span>
<span class="co">## hg19::chr1:100232177..100232198,-;hg_13495.1        15         9        26</span>
<span class="co">## hg19::chr1:100315613..100315691,+;hg_4187.1         95       109       110</span>
<span class="co">## hg19::chr1:100435545..100435597,+;hg_4201.1        141       129       101</span>
<span class="co">##                                              12hr_rep1 12hr_rep2 12hr_rep3</span>
<span class="co">## hg19::chr1:10003372..10003465,-;hg_10258.1          54        50        53</span>
<span class="co">## hg19::chr1:10003486..10003551,+;hg_541.1            35        30        40</span>
<span class="co">## hg19::chr1:100111580..100111773,+;hg_4181.1          1         2         3</span>
<span class="co">## hg19::chr1:100232177..100232198,-;hg_13495.1        20        16        13</span>
<span class="co">## hg19::chr1:100315613..100315691,+;hg_4187.1        112        94       103</span>
<span class="co">## hg19::chr1:100435545..100435597,+;hg_4201.1        132       106       125</span>
<span class="co">##                                              24hr_rep1 24hr_rep2 24hr_rep3</span>
<span class="co">## hg19::chr1:10003372..10003465,-;hg_10258.1          11        12        12</span>
<span class="co">## hg19::chr1:10003486..10003551,+;hg_541.1            22        34        50</span>
<span class="co">## hg19::chr1:100111580..100111773,+;hg_4181.1          0         1         1</span>
<span class="co">## hg19::chr1:100232177..100232198,-;hg_13495.1         7        13        10</span>
<span class="co">## hg19::chr1:100315613..100315691,+;hg_4187.1         43        74        89</span>
<span class="co">## hg19::chr1:100435545..100435597,+;hg_4201.1         84       100       121</span></code></pre>
<p>First we need to construct a design matrix describing our experiment design.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">design</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span>
  data <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">0</span>, <span class="fl">0</span>,
           <span class="fl">1</span>, <span class="fl">0</span>, <span class="fl">0</span>,
           <span class="fl">1</span>, <span class="fl">0</span>, <span class="fl">0</span>,
           <span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">0</span>,
           <span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">0</span>,
           <span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">0</span>,
           <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">1</span>,
           <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">1</span>,
           <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span>,
  ncol <span class="op">=</span> <span class="fl">3</span>,
  nrow <span class="op">=</span> <span class="fl">9</span>,
  byrow <span class="op">=</span> <span class="cn">TRUE</span>,
  dimnames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>
    <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>
      <span class="st">"00hr_rep1"</span>,
      <span class="st">"00hr_rep2"</span>,
      <span class="st">"00hr_rep3"</span>,
      <span class="st">"12hr_rep1"</span>,
      <span class="st">"12hr_rep2"</span>,
      <span class="st">"12hr_rep3"</span>,
      <span class="st">"24hr_rep1"</span>,
      <span class="st">"24hr_rep2"</span>,
      <span class="st">"24hr_rep3"</span>
    <span class="op">)</span>,
    <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"00hr"</span>, <span class="st">"12hr"</span>, <span class="st">"24hr"</span><span class="op">)</span>
  <span class="op">)</span>
<span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">design</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">##           00hr 12hr 24hr</span>
<span class="co">## 00hr_rep1    1    0    0</span>
<span class="co">## 00hr_rep2    1    0    0</span>
<span class="co">## 00hr_rep3    1    0    0</span>
<span class="co">## 12hr_rep1    0    1    0</span>
<span class="co">## 12hr_rep2    0    1    0</span>
<span class="co">## 12hr_rep3    0    1    0</span>
<span class="co">## 24hr_rep1    0    0    1</span>
<span class="co">## 24hr_rep2    0    0    1</span>
<span class="co">## 24hr_rep3    0    0    1</span></code></pre>
<p>Next, we need to preprocess the counts using <code>prepareCountsForRegression</code> function. Here, CAGE expression tags for each sample are filtered for lowly expressed promoters, normalized for the library size and transformed into counts per million (CPM). Finally, CPM are log2 transformed with addition of pseudo count 1. Moreover, we designate the base level samples, 0 hours after treatment in our example, from which basal expression level is calculated. This basal level will be used as a reference when modeling expression changes.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">mae</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/prepareCountsForRegression.html">prepareCountsForRegression</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">rinderpest_mini</span>,
                                  design <span class="op">=</span> <span class="va">design</span>,
                                  base_lvl <span class="op">=</span> <span class="st">"00hr"</span><span class="op">)</span></code></pre></div>
<p><code>xcore</code> models the expression as a function of molecular signatures. Such signatures can be constructed eg. from the known transcription factor binding sites (see <a href="#constructing-molecular-signatures">Constructing molecular signatures</a> section). Here, we will take advantage of pre-computed molecular signatures found in <code>xcoredata</code> package. Particularly, molecular signatures constructed from ReMap2020 against FANTOM5 annotation.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">remap_promoters</span>, package <span class="op">=</span> <span class="st">"xcoredata"</span><span class="op">)</span></code></pre></div>
<p>Molecular signature is a simple binary matrix indicating if transcription factor binding site was found or not in promoter vicinity.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">remap_promoters</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## 3 x 3 sparse Matrix of class "dgCMatrix"</span>
<span class="co">##                                            ENCSR000AHD.CTCF.MCF-7</span>
<span class="co">## hg19::chr1:10003314..10003321,-;hg_10256.1                      1</span>
<span class="co">## hg19::chr1:10003348..10003362,-;hg_10257.1                      1</span>
<span class="co">## hg19::chr1:10003372..10003465,-;hg_10258.1                      1</span>
<span class="co">##                                            ENCSR000AHF.TAF1.MCF-7</span>
<span class="co">## hg19::chr1:10003314..10003321,-;hg_10256.1                      1</span>
<span class="co">## hg19::chr1:10003348..10003362,-;hg_10257.1                      1</span>
<span class="co">## hg19::chr1:10003372..10003465,-;hg_10258.1                      1</span>
<span class="co">##                                            ENCSR000AKB.CTCF.GM12878</span>
<span class="co">## hg19::chr1:10003314..10003321,-;hg_10256.1                        .</span>
<span class="co">## hg19::chr1:10003348..10003362,-;hg_10257.1                        .</span>
<span class="co">## hg19::chr1:10003372..10003465,-;hg_10258.1                        .</span></code></pre>
<p>To add signatures to our <code>MultiAssayExperiment</code> object we can use <code>addSignatures</code> function. As you add your signatures remember to give them unique names.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">mae</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/addSignatures.html">addSignatures</a></span><span class="op">(</span><span class="va">mae</span>, remap <span class="op">=</span> <span class="va">remap_promoters</span><span class="op">)</span></code></pre></div>
<p>When we examine newly added signatures, we can see that some of them does not overlap any of the promoters. On the other side, depending on the signatures quality, we could expect to see signatures that overlap all of the promoters.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">colSums</a></span><span class="op">(</span><span class="va">mae</span><span class="op">[[</span><span class="st">"remap"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">##    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. </span>
<span class="co">##     0.0   272.2  1828.5  3008.1  5208.0 10224.0</span></code></pre>
<p>While, we do not provide detailed guidelines to signatures filtering, filtering out at least signatures overlapping no or all promoters is mandatory. Here, we filter signatures that overlap less than 5% or more than 95% of promoters using <code>filterSignatures</code> function.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">mae</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/filterSignatures.html">filterSignatures</a></span><span class="op">(</span><span class="va">mae</span>, min <span class="op">=</span> <span class="fl">0.05</span>, max <span class="op">=</span> <span class="fl">0.95</span><span class="op">)</span></code></pre></div>
<p>Finally, we can run our expression modeling using <code>modelGeneExpression</code> function. <code>modelGeneExpression</code> can take advantage of parallelization to speedup its calculations, to use it we need to register parallel backend.</p>
<p>Our modeling will inherently contain some level of randomness due to using cross validation, here we set the seed so that our results can be replicated.</p>
<p>This step is time consuming as we need to train a separate model for each of our samples. To speed up the calculations we will lower number of folds used in cross-validation procedure from 10 to 5 using <code>nfolds</code> argument, which is internally passed to <code>cv.glmnet</code> function. Using 6 cores on Intel(R) Xeon(R) CPU E5-2680 v3 calculations took around 4 minutes to compute and used up to 25 GB RAM. If needed memory requirement could be brought down by using lower number of cores.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># register parallel backend, for R package it might be not possible to use more than 2 cores in its vignette</span>
<span class="fu">doMC</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/doMC/man/registerDoMC.html" class="external-link">registerDoMC</a></span><span class="op">(</span>cores <span class="op">=</span> <span class="fl">2L</span><span class="op">)</span>

<span class="co"># set seed</span>
<span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">314159265</span><span class="op">)</span>

<span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/modelGeneExpression.html">modelGeneExpression</a></span><span class="op">(</span>
  mae <span class="op">=</span> <span class="va">mae</span>,
  xnames <span class="op">=</span> <span class="st">"remap"</span>,
  nfolds <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></code></pre></div>
<p>Output returned by <code>modelGeneExpression</code> is a list with following elements:</p>
<ul>
<li>
<code>regression_models</code> - a list holding <code>cv.glmnet</code> objects for each sample.</li>
<li>
<code>pvalues</code> - list of <code>data.frame</code>s for each sample, each holding signatures estimates, their estimated standard errors and p-values.</li>
<li>
<code>replicate_avg</code> - a matrix holding replicate average Z-scores with columns corresponding to groups in the design. This is likely the main output for most of the users.</li>
</ul>
<p>Let’s have a look at <code>replicate_avg</code> matrix.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">out</span> <span class="op">&lt;-</span> <span class="va">res</span><span class="op">$</span><span class="va">replicate_avg</span><span class="op">$</span><span class="va">remap</span>
<span class="va">ord</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">out</span><span class="op">[</span>, <span class="st">"24hr"</span><span class="op">]</span><span class="op">)</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="fu">pheatmap</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/pheatmap/man/pheatmap.html" class="external-link">pheatmap</a></span><span class="op">(</span>
  mat <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">out</span><span class="op">[</span><span class="va">ord</span>, <span class="op">]</span><span class="op">)</span>,
  cluster_cols <span class="op">=</span> <span class="cn">FALSE</span>,
  main <span class="op">=</span> <span class="st">"SLAM293 signatures Z-score"</span>
<span class="op">)</span></code></pre></div>
<p><img src="xcore_vignette_files/figure-html/unnamed-chunk-11-1.png" width="700"></p>
<p>Among others, our analysis indicates RELA and IRF9 as the top transcription factor associated with observed changes in expression. These transcription factors are known to play a role during measles infection, which is closely related to rinderpest (<a href="https://www.genome.jp/kegg-bin/show_pathway?hsa05162" class="external-link">KEGG Measles pathway</a>).</p>
<p>Moreover, Z-score sign indicate that promoters targeted by RELA tend to be downregulated, while those targeted by IRF9 tend to be upregulated.</p>
</div>
<div class="section level1">
<h1 id="constructing-molecular-signatures">Constructing molecular signatures<a class="anchor" aria-label="anchor" href="#constructing-molecular-signatures"></a>
</h1>
<p>Constructing molecular signatures is as simple as intersecting promoters annotation with TFBS. For example here we downloaded Motevo predictions of binding sites from SwissRegulon website. The file is GFF format and we can read it using <code>rtracklayer::import</code>. After we read it into R we will get a <code>GRanges</code> object. For our purposes it is required that TF/motif name is held under <code>name</code> column.</p>
<p>Next we construct our molecular signatures matrix using FANTOM5 annotation found in <code>xcoredata</code> package. This is done using <code>getInteractionMatrix</code> function, where we can specify the <code>ext</code> parameter that will control how much promoters are extended in both directions before intersecting with binding data.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># obtain binding sites</span>
<span class="va">motevo_hg38_f5</span> <span class="op">&lt;-</span> <span class="fu">rtracklayer</span><span class="fu">::</span><span class="fu">import</span><span class="op">(</span>
  con <span class="op">=</span> <span class="st">"https://swissregulon.unibas.ch/data/hg38_f5/hg38_sites_v1.gff.gz"</span>
<span class="op">)</span>

<span class="co"># TF / motif name must be stored as 'name'</span>
<span class="va">motevo_hg38_f5</span><span class="op">$</span><span class="va">name</span> <span class="op">&lt;-</span> <span class="va">motevo_hg38_f5</span><span class="op">$</span><span class="va">Motif</span>

<span class="co"># obtain promoters annotation, *promoter name must be stored as 'name'</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"promoters_f5"</span>, package <span class="op">=</span> <span class="st">"xcoredata"</span><span class="op">)</span>

<span class="co"># construct molecular signatures matrix</span>
<span class="va">molecular_signature</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getInteractionMatrix.html">getInteractionMatrix</a></span><span class="op">(</span>
  a <span class="op">=</span> <span class="va">promoters_f5</span>,
  b <span class="op">=</span> <span class="va">motevo_hg38_f5</span>,
  ext <span class="op">=</span> <span class="fl">500</span>
<span class="op">)</span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Maciej Migdał, Bogumił Kaczkowski.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.1.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
